{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "01927ff8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIXML \n",
    "from Bio.Blast import NCBIWWW\n",
    "import requests\n",
    "import re\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "from Bio import Medline\n",
    "from tqdm import tqdm\n",
    "from Bio import SearchIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "from bs4 import BeautifulSoup\n",
    "from Bio.SeqFeature import CompoundLocation\n",
    "import mysql.connector as SQLC\n",
    "import mysql.connector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "08f33c2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import mysql.connector as SQLC\n",
    "class DB:\n",
    "#database connection and extraction of sequences and id_genebanks\n",
    "    def __init__(self, password):        \n",
    "        if password != \"20221207\":\n",
    "            raise Exception(\"The password is wrong!\")\n",
    "\n",
    "        else:\n",
    "            self.cnx = SQLC.connect(\n",
    "                    host =\"geo.di.uminho.pt\",\n",
    "                    user =\"bioinformatica\",\n",
    "                    password = password, #\"20221207\"\n",
    "                    database =\"AP_db_KRG\")\n",
    "            self.sequences = []\n",
    "            self.ids = []\n",
    "    def extract_sequences(self):\n",
    "        cursor = self.cnx.cursor()\n",
    "        cursor.execute(\"SELECT sequence FROM Gene\")\n",
    "        self.sequences = cursor.fetchall()\n",
    "        cursor.close()\n",
    "        return self.sequences\n",
    "        \n",
    "    def extract_ids(self):\n",
    "        cursor = self.cnx.cursor()\n",
    "        cursor.execute(\"SELECT ID_genebank FROM Gene\")\n",
    "        self.ids = cursor.fetchall()\n",
    "        cursor.close()\n",
    "        return self.ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "e4a6a6ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Escolha o tamanho máximo das seq: 50\n",
      "[\"('AAADK0043071.1', 21)\", \"('AAADS0003348.1', 20)\", \"('AAADW0009410.1', 21)\", \"('AAADX0043451.1', 21)\", \"('ABAAM0346030.1', 20)\"]\n"
     ]
    }
   ],
   "source": [
    "DataBase = SQLC.connect(\n",
    "    host =\"geo.di.uminho.pt\",\n",
    "    user =\"bioinformatica\",\n",
    "    password =\"20221207\",\n",
    "    database =\"AP_db_KRG\"\n",
    ")\n",
    "DataBase.autocommit = True \n",
    "Cursor = DataBase.cursor()\n",
    "\n",
    "num= input(\"Escolha o tamanho máximo das seq: \")\n",
    "Seq_and_length=[]\n",
    "just_seq=[]\n",
    "try:\n",
    "    sql_Seq_length= f\"select Gene.ID_genebank, Gene.length from Gene where Gene.length <{num}\"\n",
    "    Cursor.execute(sql_Seq_length)\n",
    "    for row in Cursor:\n",
    "        Seq_and_length.append(str(row)) \n",
    "    \n",
    "    sql_Seq= f\"select Gene.ID_genebank from Gene where Gene.length <{num}\"\n",
    "    Cursor.execute(sql_Seq)\n",
    "    for row in Cursor:\n",
    "        just_seq.append(str(row))\n",
    "        \n",
    "except mysql.connector.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "    \n",
    "print (Seq_and_length)\n",
    "\n",
    "if just_seq == []:\n",
    "        print()\n",
    "        print('nenhuma seq inferior a esse tamanho, pesquise de novo.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "9f33dd74",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "escolha a primeira seq com que quer trabalhar: 3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\guilh\\anaconda3\\lib\\site-packages\\Bio\\Entrez\\__init__.py:658: UserWarning: \n",
      "Email address is not specified.\n",
      "\n",
      "To make use of NCBI's E-utilities, NCBI requires you to specify your\n",
      "email address with each request.  As an example, if your email address\n",
      "is A.N.Other@example.com, you can specify it as follows:\n",
      "   from Bio import Entrez\n",
      "   Entrez.email = 'A.N.Other@example.com'\n",
      "In case of excessive usage of the E-utilities, NCBI will attempt to contact\n",
      "a user at the email address provided before blocking access to the\n",
      "E-utilities.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tipo de molécula com que está a trabalhar: DNA\n",
      "[Seq('GAGTGGAGATGGCGGAGCTGT'), Seq('CAACACGGGAAACCTCACCC'), Seq('GGGATGGAAGATGGCGGCCTA'), Seq('GGTGGCTAGGGCCGCGCTGGT'), Seq('GTTTGGTGAACTTCTGCCTA')]\n",
      "GGTGGCTAGGGCCGCGCTGGT\n"
     ]
    }
   ],
   "source": [
    "list_seq=[]\n",
    "choose= input('escolha a primeira seq com que quer trabalhar: ')\n",
    "inter= int(choose)\n",
    "extr= ', '.join(just_seq)\n",
    "h= extr.replace(\"(\",'')\n",
    "hh= h.replace(\",)\",'')\n",
    "ID_al= hh.split(', ')\n",
    "select= ID_al[inter]\n",
    "list_seq.append(select)\n",
    "\n",
    "# choose2= input('escolha a segunda seq com que quer trabalhar: ')\n",
    "# inter2= int(choose2)\n",
    "# extr2= ', '.join(just_seq)\n",
    "# h2= extr2.replace(\"(\",'')\n",
    "# hh2= h2.replace(\",)\",'')\n",
    "# ID_al2= hh2.split(', ')\n",
    "# select2= ID_al2[inter2]\n",
    "# list_seq.append(select2)\n",
    "\n",
    "limpar_aspas_all= ', '.join(ID_al)\n",
    "retirar_all= limpar_aspas_all.replace(\"'\",\"\")\n",
    "otput_final_all= retirar_all.split(', ')\n",
    "\n",
    "limpar_aspas= ', '.join(list_seq)\n",
    "retirar= limpar_aspas.replace(\"'\",\"\")\n",
    "otput_final= retirar.split(', ')\n",
    "\n",
    "\n",
    "#cria lista com todos as seq com o tamanho máximo definido:\n",
    "todas_as_seq=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= otput_final_all\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "for info in records:\n",
    "    seqs= (info.seq)\n",
    "    todas_as_seq.append(seqs)\n",
    "    \n",
    "choose2= input('tipo de molécula com que está a trabalhar: ')    \n",
    "#cria lista só com as duas seq escolhidas para o alinhamento \n",
    "seq_for_al=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist1= otput_final[0]\n",
    "# idlist2= otput_final[1]\n",
    "handle1 = Entrez.efetch(db=database, id=idlist1, rettype=\"gb\") \n",
    "records1 = SeqIO.read(handle1,\"gb\")\n",
    "handle1.close()\n",
    "s1_=records1.seq\n",
    "# handle2 = Entrez.efetch(db=database, id=idlist2, rettype=\"gb\") \n",
    "# records2 = SeqIO.read(handle2,\"gb\")\n",
    "# handle2.close()\n",
    "# s2_=records2.seq\n",
    "print(todas_as_seq)\n",
    "print(s1_)\n",
    "# print(s2_)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7979ffe5",
   "metadata": {},
   "source": [
    "__Goal__ Implementation of sequence manipulation algorithms: transcription, translation, ORFs, extraction of all potential proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e37ecbb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#function that allows us to associate a codon with an amino acid- Human\n",
    "def translate_codon(cod):\n",
    "    #Translates a codon into an aminoacid using an  dictionary with the standard genetic code for Homo Sapiens.\n",
    "    tc = {\"GCT\":\"A\", \"GCC\":\"A\", \"GCA\":\"A\", \"GCG\":\"A\", \n",
    "      \"TGT\":\"C\", \"TGC\":\"C\",\n",
    "      \"GAT\":\"D\", \"GAC\":\"D\",\n",
    "      \"GAA\":\"E\", \"GAG\":\"E\",\n",
    "      \"TTT\":\"F\", \"TTC\":\"F\",\n",
    "      \"GGT\":\"G\", \"GGC\":\"G\", \"GGA\":\"G\", \"GGG\":\"G\",\n",
    "      \"CAT\":\"H\", \"CAC\":\"H\",\n",
    "      \"ATA\":\"I\", \"ATT\":\"I\", \"ATC\":\"I\",\n",
    "      \"AAA\":\"K\", \"AAG\":\"K\",\n",
    "      \"TTA\":\"L\", \"TTG\":\"L\", \"CTT\":\"L\", \"CTC\":\"L\", \"CTA\":\"L\", \"CTG\":\"L\",\n",
    "      \"ATG\":\"M\", \"AAT\":\"N\", \"AAC\":\"N\",\n",
    "      \"CCT\":\"P\", \"CCC\":\"P\", \"CCA\":\"P\", \"CCG\":\"P\",\n",
    "      \"CAA\":\"Q\", \"CAG\":\"Q\",\n",
    "      \"CGT\":\"R\", \"CGC\":\"R\", \"CGA\":\"R\", \"CGG\":\"R\", \"AGA\":\"R\", \"AGG\":\"R\",\n",
    "      \"TCT\":\"S\", \"TCC\":\"S\", \"TCA\":\"S\", \"TCG\":\"S\", \"AGT\":\"S\", \"AGC\":\"S\",\n",
    "      \"ACT\":\"T\", \"ACC\":\"T\", \"ACA\":\"T\", \"ACG\":\"T\",\n",
    "      \"GTT\":\"V\", \"GTC\":\"V\", \"GTA\":\"V\", \"GTG\":\"V\",\n",
    "      \"TGG\":\"W\",\n",
    "      \"TAT\":\"Y\", \"TAC\":\"Y\",\n",
    "      \"TAA\":\"_\", \"TAG\":\"_\", \"TGA\":\"_\"}\n",
    "    if cod in tc: return tc[cod] \n",
    "    else: return \"This codon does not exist\"\n",
    "#translate_codon(\"xbt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "09a6c7fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "You prefer use a personal query or a sequence from Data Base (1-QUERY, 2-SEQUENCE): 3\n",
      "Error: Choose an identificator present in the list\n",
      "You prefer use a personal query or a sequence from Data Base (1-QUERY, 2-SEQUENCE): 2\n",
      "[\"('AAADK0043071.1', 21)\", \"('AAADS0003348.1', 20)\", \"('AAADW0009410.1', 21)\", \"('AAADX0043451.1', 21)\", \"('ABAAM0346030.1', 20)\"]\n",
      "choose the numer that correspond to the DB sequence you want to add: 3\n",
      "GGTGGCTAGGGCCGCGCTGGT\n",
      "Type of molecule (DNA,RNA,PROTEIN): dna\n",
      "Choose a function between 1-10: 4\n",
      "[('G', 11), ('C', 5), ('T', 4), ('A', 1)]\n"
     ]
    }
   ],
   "source": [
    "while True:\n",
    "    escolha_de_input= input(\"You prefer use a personal query or a sequence from Data Base (1-QUERY, 2-SEQUENCE): \")\n",
    "    if not escolha_de_input.isdigit():\n",
    "        print(\"Error: Choose a valid identificator\")\n",
    "        continue\n",
    "    if escolha_de_input != '1' and escolha_de_input != '2':\n",
    "        print(\"Error: Choose an identificator present in the list\")\n",
    "\n",
    "        continue\n",
    "    else:\n",
    "        choose_seq_info = escolha_de_input\n",
    "        break\n",
    "if choose_seq_info == \"1\":\n",
    "    Seq_G= input('Sequence: ')\n",
    "    Sequence_type= input(\"Type of molecule (DNA,RNA,PROTEIN): \")\n",
    "    M=Sequence_type.upper()\n",
    "    RE=sequence_info(Seq_G,M)\n",
    "    dic={\"1\": RE.__len__(), \"2\":RE.letters(), \"3\":RE.validate(), \"4\":RE.contagem(), \"5\":RE.reverse_complement(),\"6\":RE.transcription(), \"7\":RE.translate(),\"8\":RE.get_orfs_DNA(),\"9\":RE.get_orfs_Proteins,\"10\":RE.protsInAA}\n",
    "    while True:\n",
    "        f=input(\"Choose a function between 1-10: \")  \n",
    "        if not a.isdigit():\n",
    "            print(\"Error: Choose a valid identificator\")\n",
    "            continue\n",
    "        if f != '1' and f != '2' and f != '3' and f != '4' and f != '5' and f != '6' and f != '7' and f != '8' and f != '9' and f != '10':\n",
    "            print(\"Error: Choose an identificator present in the list\")\n",
    "            continue\n",
    "        else:\n",
    "            result = f\n",
    "            break\n",
    "    b=dic[result]\n",
    "    print(b)\n",
    "\n",
    "elif choose_seq_info ==\"2\":\n",
    "    print (Seq_and_length)\n",
    "    seq3= int(input('choose the numer that correspond to the DB sequence you want to add: '))\n",
    "    s3=todas_as_seq[seq3]\n",
    "    print(s3)\n",
    "    Sequence_type= input(\"Type of molecule (DNA,RNA,PROTEIN): \")\n",
    "    M=Sequence_type.upper()\n",
    "    RE=sequence_info(str(s3),M)\n",
    "    dic={\"1\": RE.__len__(), \"2\":RE.letters(), \"3\":RE.validate(), \"4\":RE.contagem(), \"5\":RE.reverse_complement(),\"6\":RE.transcription(), \"7\":RE.translate(),\"8\":RE.get_orfs_DNA(),\"9\":RE.get_orfs_Proteins,\"10\":RE.protsInAA}\n",
    "    while True:\n",
    "        f=input(\"Choose a function between 1-10: \")  \n",
    "        if not a.isdigit():\n",
    "            print(\"Error: Choose a valid identificator\")\n",
    "            continue\n",
    "        if f != '1' and f != '2' and f != '3' and f != '4' and f != '5' and f != '6' and f != '7' and f != '8' and f != '9' and f != '10':\n",
    "            print(\"Error: Choose an identificator present in the list\")\n",
    "            continue\n",
    "        else:\n",
    "            result = f\n",
    "            break\n",
    "    b=dic[result]\n",
    "    print(b)\n",
    "else: print(\"You must choose 1 or 2\")\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "10e0c799",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "You prefer use a personal 1-Query or a 2-Sequence from Data Base (Press 1 or 2): 2\n",
      "Enter password: 20221207\n",
      "\n",
      "Labels for the output [1st index, 2nd index, 3rd index]: \n",
      "[Identificator, Id of the Gene, Length sequence]\n",
      "\n",
      "\n",
      "[[0, 'A00829.1', 50], [1, 'AAADK0043071.1', 21], [2, 'AAADS0003348.1', 20], [3, 'AAADW0009410.1', 21], [4, 'AAADX0043451.1', 21], [5, 'AAVX01203288.1', 50], [6, 'ABAAM0346030.1', 20], [7, 'BAAF03073813.1', 50], [8, 'DH405380.1', 50], [9, 'DS196524.1', 50], [10, 'EC923692.1', 50], [11, 'EC966285.1', 50], [12, 'M97702.1', 50], [13, 'X65482.1', 50], [14, 'X68316.1', 50]]\n",
      "Choose the identificator (1st index of each previous list) that corresponds to the DB sequence: 0\n",
      "\n",
      "Here is the sequence from DataBase:  TACCTTGTCCGAAAAGGAAAGACAAATCAAGAAGCAAACTGCTTTGGTTG\n",
      "\n",
      "Choose a function between 1-10: 0\n",
      "Error: Choose an identificator present in the list\n",
      "\n",
      "Choose a function between 1-10: a\n",
      "Error: Choose a valid function number\n",
      "\n",
      "Choose a function between 1-10: 1\n",
      "50\n"
     ]
    }
   ],
   "source": [
    "#Karyna version\n",
    "while True:\n",
    "    escolha_de_input= input(\"You prefer use a personal 1-Query or a 2-Sequence from Data Base (Press 1 or 2): \")\n",
    "    if not escolha_de_input.isdigit():\n",
    "        print(\"Error: Write just the number\")\n",
    "        continue\n",
    "    if escolha_de_input != '1' and escolha_de_input != '2':\n",
    "        print(\"Error: Choose an identificator present in the list (1 or 2)\")\n",
    "        continue\n",
    "    else:\n",
    "        choose_seq_info = escolha_de_input\n",
    "        break\n",
    "if choose_seq_info == \"1\":\n",
    "    Seq_G= input('Sequence: ')\n",
    "    Sequence_type= input(\"Type of molecule (DNA, RNA, PROTEIN): \")\n",
    "    M=Sequence_type.upper()\n",
    "    RE=sequence_info(Seq_G,M)\n",
    "    dic={\"1\": RE.__len__(), \"2\":RE.letters(), \"3\":RE.validate(), \"4\":RE.contagem(), \"5\":RE.reverse_complement(),\"6\":RE.transcription(), \"7\":RE.translate(),\"8\":RE.get_orfs_DNA(),\"9\":RE.get_orfs_Proteins,\"10\":RE.protsInAA}\n",
    "    while True:\n",
    "        print(\"1-length | 2-letter...\")\n",
    "        f=input(\"Choose a function between 1-10: \")  \n",
    "        if not f.isdigit():\n",
    "            print(\"Error: Choose a valid function number\")\n",
    "            continue\n",
    "        if int(f) not in range(1,11):\n",
    "#         if f != '1' and f != '2' and f != '3' and f != '4' and f != '5' and f != '6' and f != '7' and f != '8' and f != '9' and f != '10':\n",
    "            print(\"Error: Choose an identificator present in the list\")\n",
    "            continue\n",
    "        else:\n",
    "            result = f\n",
    "            break\n",
    "    b=dic[result]\n",
    "    print(b)\n",
    "\n",
    "elif choose_seq_info ==\"2\":\n",
    "#     print (Seq_and_length)\n",
    "    password = input(\"Enter password: \")\n",
    "    db=DB(password)\n",
    "    count = 0\n",
    "    all_info=[]\n",
    "    for index, value in enumerate(db.extract_sequences()):\n",
    "            selected_seq = []\n",
    "            selected_seq.append(count)\n",
    "            selected_seq.append(''.join(db.extract_ids()[index]))\n",
    "            selected_seq.append(len(''.join(value)))\n",
    "            all_info.append(selected_seq)\n",
    "            count += 1\n",
    "    print(\"\\nLabels for the output [1st index, 2nd index, 3rd index]: \\n[Identificator, Id of the Gene, Length sequence]\\n\\n\")\n",
    "    print(all_info)\n",
    "    identificator = [str(i[0]) for i in all_info]\n",
    "    while True:\n",
    "        seq3= input('Choose the identificator (1st index of each previous list) that corresponds to the DB sequence: ')\n",
    "        if seq3 in identificator:\n",
    "#             seq3_int=int(seq3)\n",
    "            s3=''.join(db.extract_sequences()[int(seq3)])\n",
    "            print(\"\\nHere is the sequence from DataBase: \", s3)\n",
    "#             Sequence_type= input(\"Type of molecule (DNA,RNA,PROTEIN): \")\n",
    "            Sequence_type= \"DNA\"\n",
    "#             M=Sequence_type.upper()\n",
    "            RE=sequence_info(str(s3),Sequence_type)\n",
    "            dic={\"1\": RE.__len__(), \"2\":RE.letters(), \"3\":RE.validate(), \"4\":RE.contagem(), \"5\":RE.reverse_complement(),\"6\":RE.transcription(), \"7\":RE.translate(),\"8\":RE.get_orfs_DNA(),\"9\":RE.get_orfs_Proteins,\"10\":RE.protsInAA}\n",
    "            break\n",
    "        else:        \n",
    "            print(\"Error: Choose a valid identificator\")\n",
    "            continue\n",
    "    while True:\n",
    "        f=input(\"\\nChoose a function between 1-10: \")  \n",
    "        if not f.isdigit():\n",
    "            print(\"Error: Choose a valid function number\")\n",
    "            continue\n",
    "        if int(f) not in range(1,11):\n",
    "            print(\"Error: Choose an identificator present in the list\")\n",
    "            continue\n",
    "        else:\n",
    "            result = f\n",
    "            break\n",
    "    b=dic[result]\n",
    "    print(b)\n",
    "else: print(\"You must choose 1 or 2\")\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "afd327cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#a class that allows to accomplish the objective\n",
    "class sequence_info:\n",
    "    #\"building block\", where you always have to have the sequence you are analyzing and the type of molecule: DNA, RNA, Protein\n",
    "    def __init__(self,sequence, sequence_type):\n",
    "        if sequence_type not in [\"DNA\",\"RNA\",\"PROTEIN\"]:\n",
    "            raise ValueError(\"Invalid sequence type\")\n",
    "        self.sequence=sequence.upper() \n",
    "        self.sequence_type=sequence_type\n",
    "    def __len__(self):\n",
    "        return len(self.sequence)\n",
    "    def __getitem__(self,n):\n",
    "        return self.sequence[n]\n",
    "    def __getslice__(self,i,j):\n",
    "        return self.sequence[i:j]\n",
    "    def __str__(self):\n",
    "        return self.sequence\n",
    "    def sequence_type(self):\n",
    "        return self.sequence_type\n",
    "    def letters(self):\n",
    "        #each letter possible to have for DNA, RNA , PROTEIN for later to be able to validate the information about the type of molecule corresponding to the sequence\n",
    "        if (self.sequence_type==\"DNA\"): return \"ACGT\"\n",
    "        elif(self.sequence_type==\"RNA\"):return \"ACGU\"\n",
    "        elif(self.sequence_type==\"PROTEIN\"): return \"ACDEFGHIKLMNPQRSTVWY\"\n",
    "        else: raise ValueError(\"ERROR\")\n",
    "    def validate(self): #validates information about the type of molecule associated with the sequence\n",
    "        letras=self.letters()\n",
    "        res=True\n",
    "        i=0\n",
    "        while i<len(self.sequence) and res:\n",
    "            if self.sequence[i] not in letras: res = False\n",
    "            else: i +=1\n",
    "        if res==False:\n",
    "            raise ValueError(\"Put the right sequence_type\")\n",
    "        else: return res \n",
    "    def contagem(self):\n",
    "        contagem={}\n",
    "        for x in self.sequence:\n",
    "            contagem[x]=contagem.get(x,0)+1\n",
    "        return sorted(contagem.items(), key=lambda x: x[1], reverse=True)      \n",
    "    def reverse_complement(self):#falta comprovar isto\n",
    "        if (self.validate() == False) or (self.sequence_type in [\"RNA\",\"PROTEIN\"]) :return \"only want to do reverse complement with DNA\"\n",
    "        t=self.sequence.replace(\"T\",\"a\")\n",
    "        g=t.replace(\"G\",\"c\")\n",
    "        c=g.replace(\"C\",\"g\")\n",
    "        a=c.replace(\"A\",\"t\")\n",
    "        reverse_comp=a[::-1]\n",
    "        return reverse_comp.upper()    \n",
    "    def transcription(self):\n",
    "        if (self.validate() == False) or (self.sequence_type != \"DNA\"):return \"only want to do transcription with DNA\"\n",
    "        else: return  sequence_info(self.sequence.replace(\"T\",\"U\"),\"RNA\")\n",
    "    def translate(self):\n",
    "        if (self.validate() == False) or (self.sequence_type != \"DNA\"): return \"only want to do translation with DNA\"\n",
    "        sequence_aa=\"\"\n",
    "        for x in range(0, len(self.sequence)-2,3):#3 by 3\n",
    "            codon=self.sequence[x:x+3]\n",
    "            sequence_aa += translate_codon(codon)\n",
    "        return sequence_aa\n",
    "    def get_orfs_DNA(self): #6 possible orfs\n",
    "        if (self.validate() == False) or (self.sequence_type != \"DNA\"): return \"only want to do DNA orf\"\n",
    "        inverse_sequence=self.sequence[::-1]\n",
    "        return \",\".join([self.sequence[p:]for p in range(3)] + [inverse_sequence[p:] for p in range(3)]) \n",
    "    def get_orfs_Proteins(self): # translation of the 6 possible dna orfs\n",
    "        if (self.validate() == False) or (self.sequence_type != \"DNA\"): return \"just work from DNA orf\"\n",
    "        inverse_sequence=self.sequence[::-1]\n",
    "        orfs_protein=[]\n",
    "        for p in range(3):\n",
    "            orfs_protein.append(sequence_info(self.sequence[p:],self.sequence_type).translate())\n",
    "            orfs_protein.append(sequence_info(inverse_sequence[p:],self.sequence_type).translate())\n",
    "        return \",\".join(orfs_protein)\n",
    "    def protsInAA(self):#gives all possible proteins of an amino acid sequence( from M to _)\n",
    "        seq_aa = self.translate()\n",
    "        prots_act = []\n",
    "        prots = []\n",
    "        for aa in seq_aa:\n",
    "            if aa == '_':\n",
    "                if prots_act:\n",
    "                    for p in prots_act:\n",
    "                        prots.append(p)\n",
    "                    prots_act = []\n",
    "            else:\n",
    "                if aa == 'M':\n",
    "                    prots_act.append('')\n",
    "                for i in range(len(prots_act)):\n",
    "                    prots_act[i] += aa\n",
    "        return \",\".join(prots) \n",
    " #if name=main é para conseguir importar direitinho os nodulos ou seja quanto estou neste ficheiro o name =main e ele faz a classe se for para outro ficheiro e importar a classe ele faz na mesma a classe sem repetir tufo o que aqui tenho\n",
    "# https://www.youtube.com/watch?v=3dHyS1W4TIE\n",
    "#if __name__ == \"__main__\":\n",
    "    #s1_=\"AA??\"\n",
    "    s1 = sequence_info(\"ACTGCTG\",\"PROTEIN\")\n",
    "    \n",
    "    #s2 = sequence_info(\"MKVVLSVQERSVVSLL_\", \"PROTEIN\")\n",
    "    #s3 = sequence_info(\"ATTTTBTT\",\"PROTEIN\")\n",
    "    #print(\"sequence 1 is {}? {}\".format(s1.sequence_type,s1.validate()))\n",
    "    s3 = s1.validate()\n",
    "    print(s3)\n",
    "    #s4 = s1.reverse_complement()\n",
    "    #print(s4)\n",
    "#     s5= s1.translate()\n",
    "#     print(s5)\n",
    "    #s6=s1.get_orfs_DNA()\n",
    "    #print(s6)\n",
    "    #s8=s1.get_orfs_Proteins()\n",
    "    #print(s8)\n",
    "#     s9=s1.protsInAA()\n",
    "#     print(s9)\n",
    "#     s10=s1.contagem()\n",
    "#     print(s10)\n",
    "    #s11=s1.__len__()\n",
    "    #print(s11)\n",
    "    #s12=s1.letters()\n",
    "    #print(s12)\n",
    "    \n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "fa3d3a44",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "...............\n",
      "----------------------------------------------------------------------\n",
      "Ran 15 tests in 0.012s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "import unittest\n",
    "class TestSequenceInfo(unittest.TestCase):\n",
    "    def test_init(self):\n",
    "        seq = sequence_info(\"AGCT\", \"DNA\")\n",
    "        self.assertEqual(seq.sequence, \"AGCT\")\n",
    "        self.assertEqual(seq.sequence_type, \"DNA\")\n",
    "        \n",
    "        seq=sequence_info(\"AGCU\",\"RNA\")\n",
    "        self.assertEqual(seq.sequence, \"AGCU\")\n",
    "        self.assertEqual(seq.sequence_type, \"RNA\")\n",
    "        \n",
    "        with self.assertRaises(ValueError) as error:\n",
    "            seq = sequence_info(\"AA??\", \"DNS\")\n",
    "        self.assertEqual(error.exception.args[0],\"Invalid sequence type\")\n",
    "        \n",
    "        seq = sequence_info(\"AA??\", \"DNA\")\n",
    "        self.assertEqual(seq.sequence, \"AA??\")\n",
    "        self.assertEqual(seq.sequence_type, \"DNA\")\n",
    "\n",
    "    def test_len(self):\n",
    "        seq = sequence_info(\"AGCT\", \"DNA\")\n",
    "        self.assertEqual(len(seq), 4)\n",
    "        \n",
    "        seq = sequence_info(\"AGCU\",\"RNA\")\n",
    "        self.assertEqual(len(seq), 4)\n",
    "        \n",
    "        seq=sequence_info(\"AA??\",\"DNA\")\n",
    "        self.assertEqual(len(seq),4)\n",
    "\n",
    "    def test_getitem(self):\n",
    "        seq = sequence_info(\"AGCT\", \"DNA\")\n",
    "        self.assertEqual(seq[0], \"A\")\n",
    "        \n",
    "        seq = sequence_info(\"AGCU\",\"RNA\")\n",
    "        self.assertEqual(seq[0], \"A\")\n",
    "        \n",
    "        seq=sequence_info(\"AA??\",\"DNA\")\n",
    "        self.assertEqual(seq[0],\"A\")\n",
    "        \n",
    "    def test_getslice(self):\n",
    "        seq = sequence_info(\"AGCT\", \"DNA\")\n",
    "        self.assertEqual(seq[0:2], \"AG\")\n",
    "        \n",
    "        seq = sequence_info(\"AGCU\",\"RNA\")\n",
    "        self.assertEqual(seq[0:2], \"AG\")\n",
    "        \n",
    "        seq=sequence_info(\"AA??\",\"DNA\")\n",
    "        self.assertEqual(seq[1:3],\"A?\")\n",
    "\n",
    "    def test_str(self):\n",
    "        seq = sequence_info(\"AGCT\", \"DNA\")\n",
    "        self.assertEqual(str(seq), \"AGCT\")\n",
    "        \n",
    "        seq = sequence_info(\"AGCU\",\"RNA\")\n",
    "        self.assertEqual(str(seq), \"AGCU\")\n",
    "        \n",
    "        seq=sequence_info(\"AA??\",\"DNA\")\n",
    "        self.assertEqual(str(seq), \"AA??\")\n",
    "\n",
    "    def test_sequence_type(self):\n",
    "        seq = sequence_info(\"AGCT\", \"DNA\")\n",
    "        self.assertEqual(seq.sequence_type, \"DNA\")\n",
    "        \n",
    "        seq=sequence_info(\"AGCU\",\"RNA\")\n",
    "        self.assertEqual(seq.sequence_type, \"RNA\")\n",
    "        \n",
    "        seq=sequence_info(\"AA??\",\"DNA\")\n",
    "        self.assertEqual(seq.sequence_type, \"DNA\")\n",
    "        \n",
    "    def test_letters(self):\n",
    "        seq = sequence_info(\"AGCT\", \"DNA\")\n",
    "        self.assertEqual(seq.letters(), \"ACGT\")\n",
    "        \n",
    "        seq=sequence_info(\"AGCU\",\"RNA\")\n",
    "        self.assertEqual(seq.letters(), \"ACGU\")\n",
    "        \n",
    "        seq = sequence_info(\"AA??\", \"DNA\")\n",
    "        self.assertEqual(seq.letters(), \"ACGT\")\n",
    "\n",
    "    def test_validate(self):\n",
    "        seq = sequence_info(\"AGCT\", \"DNA\")\n",
    "        self.assertTrue(seq.validate())\n",
    "        \n",
    "        seq = sequence_info(\"AGCU\",\"RNA\")\n",
    "        self.assertTrue(seq.validate())\n",
    "        \n",
    "        seq = sequence_info(\"AA??\", \"DNA\")\n",
    "        try:\n",
    "            seq.validate()\n",
    "        except ValueError as ve:\n",
    "            assert str(ve) == \"Put the right sequence_type\", f\"Unexpected error message: {str(ve)}\"\n",
    "        else:\n",
    "            assert False, \"ValueError not raised\"\n",
    "\n",
    "    def test_contagem(self):\n",
    "        seq = sequence_info(\"AGCT\", \"DNA\")\n",
    "        self.assertEqual(seq.contagem(), [(\"A\", 1), (\"G\", 1), (\"C\", 1), (\"T\", 1)])\n",
    "        \n",
    "        seq = sequence_info(\"AGCU\", \"RNA\")\n",
    "        self.assertEqual(seq.contagem(), [(\"A\", 1), (\"G\", 1), (\"C\", 1), (\"U\", 1)])\n",
    "        \n",
    "        seq= sequence_info(\"AA??\",\"DNA\")\n",
    "        self.assertEqual(seq.contagem(),[(\"A\",2),(\"?\",2)])\n",
    "\n",
    "    def test_reverse_complement(self):\n",
    "        seq = sequence_info(\"AGCT\", \"DNA\")\n",
    "        self.assertEqual(seq.reverse_complement(), \"AGCT\")\n",
    "        \n",
    "        seq = sequence_info(\"AGCU\", \"RNA\")\n",
    "        self.assertEqual(seq.reverse_complement(), \"only want to do reverse complement with DNA\")\n",
    "        \n",
    "        seq = sequence_info(\"AA??\", \"DNA\")\n",
    "        try:\n",
    "            seq.reverse_complement()\n",
    "        except ValueError as ve:\n",
    "            assert str(ve) == \"Put the right sequence_type\", f\"Unexpected error message: {str(ve)}\"\n",
    "        else:\n",
    "            assert False, \"ValueError not raised\"\n",
    "\n",
    "    def test_transcription(self):\n",
    "        seq = sequence_info(\"AGCT\", \"DNA\")\n",
    "        self.assertEqual(seq.transcription().sequence, \"AGCU\")\n",
    "        seq = sequence_info(\"AGCU\", \"RNA\")\n",
    "        self.assertEqual(seq.transcription(), \"only want to do transcription with DNA\")\n",
    "        \n",
    "        seq = sequence_info(\"AA??\", \"DNA\")\n",
    "        try:\n",
    "            seq.transcription()\n",
    "        except ValueError as ve:\n",
    "            assert str(ve) == \"Put the right sequence_type\", f\"Unexpected error message: {str(ve)}\"\n",
    "        else:\n",
    "            assert False, \"ValueError not raised\"\n",
    "\n",
    "    def test_translate(self):\n",
    "        seq = sequence_info(\"AGCT\", \"DNA\")\n",
    "        self.assertEqual(seq.translate(), \"S\")\n",
    "        \n",
    "        seq = sequence_info(\"AGCU\", \"RNA\")\n",
    "        self.assertEqual(seq.translate(), \"only want to do translation with DNA\")\n",
    "        \n",
    "        seq = sequence_info(\"AA??\", \"DNA\")\n",
    "        try:\n",
    "            seq.translate()\n",
    "        except ValueError as ve:\n",
    "            assert str(ve) == \"Put the right sequence_type\", f\"Unexpected error message: {str(ve)}\"\n",
    "        else:\n",
    "            assert False, \"ValueError not raised\"\n",
    "\n",
    "    def test_get_orfs_DNA(self):\n",
    "        seq = sequence_info(\"AGCT\", \"DNA\")\n",
    "        self.assertEqual(seq.get_orfs_DNA(), \"AGCT,GCT,CT,TCGA,CGA,GA\")\n",
    "        \n",
    "        seq = sequence_info(\"AGCU\", \"RNA\")\n",
    "        self.assertEqual(seq.get_orfs_DNA(), \"only want to do DNA orf\")\n",
    "        \n",
    "        seq = sequence_info(\"AA??\", \"DNA\")\n",
    "        try:\n",
    "            seq.get_orfs_DNA()\n",
    "        except ValueError as ve:\n",
    "            assert str(ve) == \"Put the right sequence_type\", f\"Unexpected error message: {str(ve)}\"\n",
    "        else:\n",
    "            assert False, \"ValueError not raised\"\n",
    "\n",
    "    def test_get_orfs_Proteins(self):\n",
    "        seq = sequence_info(\"AGCT\", \"DNA\")\n",
    "        self.assertEqual(seq.get_orfs_Proteins(), \"S,S,A,R,,\")\n",
    "        \n",
    "        seq = sequence_info(\"AGCU\", \"RNA\")\n",
    "        self.assertEqual(seq.get_orfs_Proteins(), \"just work from DNA orf\")\n",
    "        \n",
    "        seq = sequence_info(\"AA??\", \"DNA\")\n",
    "        try:\n",
    "            seq.get_orfs_Proteins()\n",
    "        except ValueError as ve:\n",
    "            assert str(ve) == \"Put the right sequence_type\", f\"Unexpected error message: {str(ve)}\"\n",
    "        else:\n",
    "            assert False, \"ValueError not raised\"\n",
    "\n",
    "    def test_protsInAA(self):\n",
    "        seq = sequence_info(\"AGCT\", \"DNA\")\n",
    "        self.assertEqual(seq.protsInAA(), \"\")\n",
    "        \n",
    "        seq = sequence_info(\"AGCU\", \"RNA\")\n",
    "        self.assertEqual(seq.protsInAA(), \"\")\n",
    "        \n",
    "        seq = sequence_info(\"AA??\", \"DNA\")\n",
    "        try:\n",
    "            seq.protsInAA()\n",
    "        except ValueError as ve:\n",
    "            assert str(ve) == \"Put the right sequence_type\", f\"Unexpected error message: {str(ve)}\"\n",
    "        else:\n",
    "            assert False, \"ValueError not raised\"\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    unittest.main(argv=['first-arg-is-ignored'], exit=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "119de4be",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
