{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "38bf8a22",
   "metadata": {},
   "source": [
    "## Ligação BD:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "83cb78e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests  \n",
    "import re\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "from Bio import Medline\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "from Bio import SearchIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "import mysql.connector as SQLC\n",
    "import mysql.connector\n",
    "from datetime import date\n",
    "import time\n",
    "from bs4 import BeautifulSoup\n",
    "from Bio.SeqFeature import CompoundLocation\n",
    "from datetime import datetime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1948ae9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import mysql.connector as SQLC\n",
    "class DB:\n",
    "#database connection and extraction of sequences and id_genebanks\n",
    "    def __init__(self, password):        \n",
    "        if password != \"20221207\":\n",
    "            raise Exception(\"The password is wrong!\")\n",
    "\n",
    "        else:\n",
    "            self.cnx = SQLC.connect(\n",
    "                    host =\"geo.di.uminho.pt\",\n",
    "                    user =\"bioinformatica\",\n",
    "                    password = password, #\"20221207\"\n",
    "                    database =\"AP_db_KRG\")\n",
    "            self.sequences = []\n",
    "            self.ids = []\n",
    "    def extract_sequences(self):\n",
    "        cursor = self.cnx.cursor()\n",
    "        cursor.execute(\"SELECT sequence FROM Gene\")\n",
    "        self.sequences = cursor.fetchall()\n",
    "        cursor.close()\n",
    "        return self.sequences\n",
    "        \n",
    "    def extract_ids(self):\n",
    "        cursor = self.cnx.cursor()\n",
    "        cursor.execute(\"SELECT ID_genebank FROM Gene\")\n",
    "        self.ids = cursor.fetchall()\n",
    "        cursor.close()\n",
    "        return self.ids\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "065aefb9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A00829.1', 'AAADK0043071.1', 'AAADS0003348.1', 'AAADW0009410.1', 'AAADX0043451.1', 'AAVX01203288.1', 'ABAAM0346030.1', 'BAAF03073813.1', 'DH405380.1', 'DS196524.1', 'EC923692.1', 'EC966285.1', 'M97702.1', 'X65482.1', 'X68316.1']\n"
     ]
    }
   ],
   "source": [
    "var = DB(\"20221207\")\n",
    "var.extract_sequences()\n",
    "var.extract_ids()\n",
    "lista=[]\n",
    "for i in var.extract_ids():\n",
    "    lista.append(''.join(i))\n",
    "print(lista)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bfaf17ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Select the maximum length of sequence (top maximum is 200): 201\n",
      "Enter password: 20221207\n",
      "Error: The maximum length must be equal or less than 200\n",
      "Select the maximum length of sequence (top maximum is 200): 199\n",
      "Enter password: 2022\n",
      "Error: wrong password\n",
      "Select the maximum length of sequence (top maximum is 200): 200\n",
      "Enter password: 20221207\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'DB' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[1;32mIn [1]\u001b[0m, in \u001b[0;36m<cell line: 11>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     14\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m result \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mInputs are valid.\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m     15\u001b[0m     all_info \u001b[38;5;241m=\u001b[39m [] \n\u001b[1;32m---> 16\u001b[0m     db\u001b[38;5;241m=\u001b[39m\u001b[43mDB\u001b[49m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m20221207\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m     17\u001b[0m     count, minimum \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m , \u001b[38;5;241m200\u001b[39m\n\u001b[0;32m     18\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m index, value \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(db\u001b[38;5;241m.\u001b[39mextract_sequences()):\n",
      "\u001b[1;31mNameError\u001b[0m: name 'DB' is not defined"
     ]
    }
   ],
   "source": [
    "def validate_inputs_align(num, password):        \n",
    "    if not num.isdigit():\n",
    "        return \"Error: The maximum length must be a number\"\n",
    "    if int(num)>200:\n",
    "        return \"Error: The maximum length must be equal or less than 200\"\n",
    "    if password != \"20221207\":\n",
    "        return \"Error: wrong password\"    \n",
    "    return \"Inputs are valid.\"\n",
    "\n",
    "while True:\n",
    "    num = input(\"Select the maximum length of sequence (top maximum is 200): \")\n",
    "    password = input(\"Enter password: \")\n",
    "    result = validate_inputs_align(num, password)\n",
    "    if result == \"Inputs are valid.\":\n",
    "        all_info = [] \n",
    "        db=DB(\"20221207\")\n",
    "        count, minimum = 0 , 200\n",
    "        for index, value in enumerate(db.extract_sequences()):\n",
    "            selected_seq = []\n",
    "            if len(''.join(value)) <= int(num):\n",
    "                selected_seq.append(count)\n",
    "                selected_seq.append(''.join(db.extract_ids()[index]))\n",
    "                selected_seq.append(len(''.join(value)))\n",
    "                all_info.append(selected_seq)\n",
    "                count += 1\n",
    "            if len(''.join(value)) < minimum:\n",
    "                minimum = len(''.join(value))\n",
    "        if all_info == []:\n",
    "            print(\"The minimum of the length is {}\".format(minimum))\n",
    "        else:\n",
    "            print(\"\\nLabels for the output: \\n[1st index is an identificator, 2nd index is the id of the Gene, and 3rd is the length sequence]\\n\\n\")\n",
    "            print(all_info)\n",
    "        break\n",
    "    else:\n",
    "        print(result)\n",
    "\n",
    "choose1= input('\\nChoose the identificator of a seq-1 that you want to work with: ')\n",
    "choose2= input('Choose the identificator of a seq-2 that you want to work with: ')\n",
    "s1_ = ''.join(db.extract_sequences()[int(choose1)])\n",
    "s2_ = ''.join(db.extract_sequences()[int(choose2)])\n",
    "print('\\n{}'.format(s1_))\n",
    "print(s2_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b3ddef37",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Select the maximum length of sequence (top maximum is 200): 50\n",
      "Enter password: 20221207\n",
      "\n",
      "Labels for the output: \n",
      "[1st index is an identificator, 2nd index is the id of the Gene, and 3rd is the length sequence]\n",
      "\n",
      "\n",
      "[[0, 'A00829.1', 50], [1, 'AAADK0043071.1', 21], [2, 'AAADS0003348.1', 20], [3, 'AAADW0009410.1', 21], [4, 'AAADX0043451.1', 21], [5, 'AAVX01203288.1', 50], [6, 'ABAAM0346030.1', 20], [7, 'BAAF03073813.1', 50], [8, 'DH405380.1', 50], [9, 'DS196524.1', 50], [10, 'EC923692.1', 50], [11, 'EC966285.1', 50], [12, 'M97702.1', 50], [13, 'X65482.1', 50], [14, 'X68316.1', 50]]\n",
      "\n",
      "Choose the identificator of a seq-1 that you want to work with: 0\n",
      "Choose the identificator of a seq-2 that you want to work with: 12\n",
      "\n",
      "TACCTTGTCCGAAAAGGAAAGACAAATCAAGAAGCAAACTGCTTTGGTTG\n",
      "GAATTCGTGCGCAACGCAATCCGGAAGTACAAGCAACTGCAGAAGCTGGT\n"
     ]
    }
   ],
   "source": [
    "#Final\n",
    "def validate_inputs_align(num, password):        \n",
    "    if not num.isdigit():\n",
    "        return \"Error: The maximum length must be a number\"\n",
    "    if int(num)>200:\n",
    "        return \"Error: The maximum length must be equal or less than 200\"\n",
    "    if password != \"20221207\":\n",
    "        return \"Error: wrong password\"    \n",
    "    return \"Inputs are valid.\"\n",
    "\n",
    "def get_sequences_info(num, password):\n",
    "    result = validate_inputs_align(num, password)\n",
    "    if result == \"Inputs are valid.\":\n",
    "        all_info = [] \n",
    "        db=DB(\"20221207\")\n",
    "        count, minimum = 0 , 200\n",
    "        for index, value in enumerate(db.extract_sequences()):\n",
    "            selected_seq = []\n",
    "            if len(''.join(value)) <= int(num):\n",
    "                selected_seq.append(count)\n",
    "                selected_seq.append(''.join(db.extract_ids()[index]))\n",
    "                selected_seq.append(len(''.join(value)))\n",
    "                all_info.append(selected_seq)\n",
    "                count += 1\n",
    "            if len(''.join(value)) < minimum:\n",
    "                minimum = len(''.join(value))\n",
    "        if all_info == []:\n",
    "            return \"The minimum of the length is {}\".format(minimum)\n",
    "        else:\n",
    "            return all_info\n",
    "    else:\n",
    "        return result\n",
    "\n",
    "while True:\n",
    "    num = input(\"Select the maximum length of sequence (top maximum is 200): \")\n",
    "    password = input(\"Enter password: \")\n",
    "    result = get_sequences_info(num, password)\n",
    "    if type(result) == str:\n",
    "        print(result)\n",
    "    else:\n",
    "        print(\"\\nLabels for the output: \\n[1st index is an identificator, 2nd index is the id of the Gene, and 3rd is the length sequence]\\n\\n\")\n",
    "        print(result)\n",
    "        break\n",
    "\n",
    "while True:\n",
    "    choose1 = input('\\nChoose the identificator of a seq-1 that you want to work with: ')\n",
    "    choose2 = input('Choose the identificator of a seq-2 that you want to work with: ')\n",
    "    if not choose1.isdigit() or not choose2.isdigit():\n",
    "        print(\"Error: Choose a valid identificator\")\n",
    "        continue\n",
    "    choose1 = int(choose1)\n",
    "    choose2 = int(choose2)\n",
    "    identificators = [i[0] for i in result]\n",
    "    if choose1 not in identificators or choose2 not in identificators:\n",
    "        print(\"Error: Choose an identificator present in the list\")\n",
    "        continue\n",
    "    else:\n",
    "        s1_ = ''.join(db.extract_sequences()[int(choose1)])\n",
    "        s2_ = ''.join(db.extract_sequences()[int(choose2)])\n",
    "        print('\\n{}'.format(s1_))\n",
    "        print(s2_)\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5551fc62",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "id": "a9c94e19",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "4b99101a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DataBase = SQLC.connect(\n",
    "#     host =\"geo.di.uminho.pt\",\n",
    "#     user =\"bioinformatica\",\n",
    "#     password =\"20221207\",\n",
    "#     database =\"AP_db_KRG\"\n",
    "# )\n",
    "# DataBase.autocommit = True \n",
    "# Cursor = DataBase.cursor()\n",
    "\n",
    "# num= input(\"Escolha o tamanho máximo das seq: \")\n",
    "# Seq_and_length=[]\n",
    "# just_seq=[]\n",
    "# try:\n",
    "#     sql_Seq_length= f\"select Gene.ID_genebank, Gene.length from Gene where Gene.length <{num}\"\n",
    "#     Cursor.execute(sql_Seq_length)\n",
    "#     for row in Cursor:\n",
    "#         Seq_and_length.append(str(row)) \n",
    "    \n",
    "#     sql_Seq= f\"select Gene.ID_genebank from Gene where Gene.length <{num}\"\n",
    "#     Cursor.execute(sql_Seq)\n",
    "#     for row in Cursor:\n",
    "#         just_seq.append(str(row))\n",
    "        \n",
    "# except mysql.connector.Error as e:\n",
    "#     print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "    \n",
    "# if just_seq == []:\n",
    "#         print()\n",
    "#         print('nenhuma seq inferior a esse tamanho, pesquise de novo.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "46ddbcc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#escolher as duas seq e criar listas com output necessários para extrair as seqs\n",
    "# print (Seq_and_length)\n",
    "# list_seq=[]\n",
    "# choose= input('escolha a primeira seq com que quer trabalhar: ')\n",
    "# inter= int(choose)\n",
    "# extr= ', '.join(just_seq)\n",
    "# h= extr.replace(\"(\",'')\n",
    "# hh= h.replace(\",)\",'')\n",
    "# ID_al= hh.split(', ')\n",
    "# select= ID_al[inter]\n",
    "# list_seq.append(select)\n",
    "\n",
    "# choose2= input('escolha a segunda seq com que quer trabalhar: ')\n",
    "# inter2= int(choose2)\n",
    "# extr2= ', '.join(just_seq)\n",
    "# h2= extr2.replace(\"(\",'')\n",
    "# hh2= h2.replace(\",)\",'')\n",
    "# ID_al2= hh2.split(', ')\n",
    "# select2= ID_al2[inter2]\n",
    "# list_seq.append(select2)\n",
    "\n",
    "# limpar_aspas_all= ', '.join(ID_al)\n",
    "# retirar_all= limpar_aspas_all.replace(\"'\",\"\")\n",
    "# otput_final_all= retirar_all.split(', ')\n",
    "\n",
    "# limpar_aspas= ', '.join(list_seq)\n",
    "# retirar= limpar_aspas.replace(\"'\",\"\")\n",
    "# otput_final= retirar.split(', ')\n",
    "\n",
    "\n",
    "# #cria lista com todos as seq com o tamanho máximo definido:\n",
    "# todas_as_seq=[]\n",
    "# database = 'nucleotide'\n",
    "# email= 'rodrigoce9@gmail.com'\n",
    "# idlist= otput_final_all\n",
    "# handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "# records = list(SeqIO.parse(handle,\"gb\"))\n",
    "# handle.close()\n",
    "# for info in records:\n",
    "#     seqs= (info.seq)\n",
    "#     todas_as_seq.append(seqs)\n",
    "    \n",
    "    \n",
    "# #cria lista só com as duas seq escolhidas para o alinhamento \n",
    "# seq_for_al=[]\n",
    "# database = 'nucleotide'\n",
    "# email= 'rodrigoce9@gmail.com'\n",
    "# idlist1= otput_final[0]\n",
    "# idlist2= otput_final[1]\n",
    "# handle1 = Entrez.efetch(db=database, id=idlist1, rettype=\"gb\") \n",
    "# records1 = SeqIO.read(handle1,\"gb\")\n",
    "# handle1.close()\n",
    "# s1_=records1.seq\n",
    "# handle2 = Entrez.efetch(db=database, id=idlist2, rettype=\"gb\") \n",
    "# records2 = SeqIO.read(handle2,\"gb\")\n",
    "# handle2.close()\n",
    "# s2_=records2.seq\n",
    "\n",
    "# print(s1_)\n",
    "# print(s2_)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8af91a19",
   "metadata": {},
   "source": [
    "### Matriz:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b66487dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import io\n",
    "\n",
    "class Mat:\n",
    "    \n",
    "    def __init__(self, rows, cols):\n",
    "        self.mat = [[0 for c in range(cols)]\n",
    "                    for r in range(rows)]\n",
    "    def numRows (self): return len(self.mat)\n",
    "    def numCols (self): return len(self.mat[0])\n",
    "    def __str__(self):\n",
    "        return '\\n'.join(' '.join(str(val) for val in row)\n",
    "                         for row in self.mat)\n",
    "    def __repr__(self):\n",
    "        return str(self)\n",
    "    def __getitem__ (self, n):\n",
    "        return self.mat[n]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6fda677",
   "metadata": {},
   "source": [
    "### NW:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "5f514e49",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    -   G   G   G   A   T   G   G   A   A   G   A   T   G   G   C   G   G   C   C   T   A\n",
      "-   0  -4  -8 -12 -16 -20 -24 -28 -32 -36 -40 -44 -48 -52 -56 -60 -64 -68 -72 -76 -80 -84\n",
      "C  -4  -1  -5  -9 -13 -17 -21 -25 -29 -33 -37 -41 -45 -49 -53 -53 -57 -61 -65 -69 -73 -77\n",
      "A  -8  -5  -2  -6  -6 -10 -14 -18 -22 -26 -30 -34 -38 -42 -46 -50 -54 -58 -62 -66 -70 -70\n",
      "A -12  -9  -6  -3  -3  -7 -11 -15 -15 -19 -23 -27 -31 -35 -39 -43 -47 -51 -55 -59 -63 -67\n",
      "C -16 -13 -10  -7  -4  -4  -8 -12 -16 -16 -20 -24 -28 -32 -36 -36 -40 -44 -48 -52 -56 -60\n",
      "A -20 -17 -14 -11  -4  -5  -5  -9  -9 -13 -17 -17 -21 -25 -29 -33 -37 -41 -45 -49 -53 -53\n",
      "C -24 -21 -18 -15  -8  -5  -6  -6 -10 -10 -14 -18 -18 -22 -26 -26 -30 -34 -38 -42 -46 -50\n",
      "G -28 -21 -18 -15 -12  -9  -2  -3  -7 -11  -7 -11 -15 -15 -19 -23 -23 -27 -31 -35 -39 -43\n",
      "G -32 -25 -18 -15 -16 -13  -6   1  -3  -7  -8  -8 -12 -12 -12 -16 -20 -20 -24 -28 -32 -36\n",
      "G -36 -29 -22 -15 -16 -17 -10  -3   0  -4  -4  -8  -9  -9  -9 -13 -13 -17 -21 -25 -29 -33\n",
      "A -40 -33 -26 -19 -12 -16 -14  -7   0   3  -1  -1  -5  -9 -10 -10 -14 -14 -18 -22 -26 -26\n",
      "A -44 -37 -30 -23 -16 -13 -17 -11  -4   3   2   2  -2  -6 -10 -11 -11 -15 -15 -19 -23 -23\n",
      "A -48 -41 -34 -27 -20 -17 -14 -15  -8  -1   2   5   1  -3  -7 -11 -12 -12 -16 -16 -20 -20\n",
      "C -52 -45 -38 -31 -24 -21 -18 -15 -12  -5  -2   1   4   0  -4  -4  -8 -12  -9 -13 -17 -21\n",
      "C -56 -49 -42 -35 -28 -25 -22 -19 -16  -9  -6  -3   0   3  -1  -1  -5  -9  -9  -6 -10 -14\n",
      "T -60 -53 -46 -39 -32 -25 -26 -23 -20 -13 -10  -7   0  -1   2  -2  -2  -6 -10 -10  -3  -7\n",
      "C -64 -57 -50 -43 -36 -29 -26 -27 -24 -17 -14 -11  -4  -1  -2   5   1  -3  -3  -7  -7  -4\n",
      "A -68 -61 -54 -47 -40 -33 -30 -27 -24 -21 -18 -11  -8  -5  -2   1   4   0  -4  -4  -8  -4\n",
      "C -72 -65 -58 -51 -44 -37 -34 -31 -28 -25 -22 -15 -12  -9  -6   1   0   3   3  -1  -5  -8\n",
      "C -76 -69 -62 -55 -48 -41 -38 -35 -32 -29 -26 -19 -16 -13 -10  -3   0  -1   6   6   2  -2\n",
      "C -80 -73 -66 -59 -52 -45 -42 -39 -36 -33 -30 -23 -20 -17 -14  -7  -4  -1   2   9   5   1\n",
      "\n",
      "  - G G G A T G G A A G A T G G C G G C C T A\n",
      "- 0 E E E E E E E E E E E E E E E E E E E E E\n",
      "C C D D D D D D D D D D D D D D D E E D D E E\n",
      "A C D D D D E E E D D E D E E E E D D D D D D\n",
      "A C D D D D D D D D D E D E E E E E E E E E D\n",
      "C C D D D D D D D D D D D D D D D E E D D E E\n",
      "A C D D D D D D D D D D D E E E E D D D D D D\n",
      "C C D D D C D D D D D D D D D D D E E D D E E\n",
      "G C D D D C D D D D D D E E D D E D D E E E E\n",
      "G C D D D D D D D E E D D D D D E D D E E E E\n",
      "G C D D D D D D D D D D E D D D D D D D D D D\n",
      "A C C C C D E C C D D E D E E D D D D D D D D\n",
      "A C C C C D D D C D D D D D D D D D D D D D D\n",
      "A C C C C D D D C D D D D D D D D D D D D D D\n",
      "C C C C C C D D D C C D D D D D D E E D D D D\n",
      "C C C C C C D D D D C D D D D D D D D D D E E\n",
      "T C C C C C D D D D C D D D D D D D D D D D E\n",
      "C C C C C C C D D D C D D C D D D E D D D C D\n",
      "A C C C C D C D D D D D D C D D C D D D D D D\n",
      "C C C C C C C D D D D D C D D D D D D D D D C\n",
      "C C C C C C C D D D D D C D D D D D D D D E E\n",
      "C C C C C C C D D D D D C D D D D D D D D D D\n",
      "\n",
      "\n",
      "CAACACGGGAAACCTC-ACC-C\n",
      "-GGGATGGAAGATGGCGGCCTA\n"
     ]
    }
   ],
   "source": [
    "class NW:\n",
    "    def __init__(self, s1, s2, gap_penalty, match_score, mismatch_penalty):\n",
    "        self.s1 = s1 \n",
    "        self.s2 = s2 \n",
    "        self.gap_penalty= gap_penalty\n",
    "        self.match_score= match_score\n",
    "        self.mismatch_penalty= mismatch_penalty\n",
    "        self.mat = Mat(len(s1) + 1, len(s2) + 1)\n",
    "        self.tr  = Mat(len(s1) + 1, len(s2) + 1)\n",
    "\n",
    "        for L in range(len(s1)):\n",
    "            self.mat[L + 1][0] = self.gap_penalty * (L + 1) \n",
    "            self.tr[L + 1][0]  = 'C' \n",
    "        for C in range(len(s2)):\n",
    "            self.mat[0][C + 1] = self.gap_penalty * (C + 1) \n",
    "            self.tr[0][C + 1]  = 'E'\n",
    "\n",
    "        for L, x1 in enumerate(s1):\n",
    "            for C, x2 in enumerate(s2):\n",
    "                possiveis = [\n",
    "                    self.mat[L  ][C    ] + (self.match_score if x1 == x2 else self.mismatch_penalty),   # Diagonal\n",
    "                    self.mat[L+1][C    ] + self.gap_penalty,               # Esquerda\n",
    "                    self.mat[L  ][C + 1] + self.gap_penalty,               # Cima\n",
    "                ]\n",
    "                dirs = \"DEC\"\n",
    "                self.mat[L + 1][C + 1] = max(possiveis)\n",
    "                self.tr[L + 1][C + 1] = dirs[possiveis.index(self.mat[L + 1][C + 1])]\n",
    "                \n",
    "    def rebuild(self):\n",
    "        L = len(self.s1)\n",
    "        C = len(self.s2)\n",
    "        S1 = \"\"\n",
    "        S2 = \"\"\n",
    "    \n",
    "        dirs = {\n",
    "            'D' : (-1, -1),\n",
    "            'E' : ( 0, -1),\n",
    "            'C' : (-1,  0)\n",
    "        }\n",
    "        while L > 0 or C > 0:\n",
    "            \n",
    "            DL, DC = dirs[self.tr[L][C]]\n",
    "            if self.tr[L][C] == \"D\":\n",
    "                S1 = self.s1[L - 1] + S1 \n",
    "                S2 = self.s2[C - 1] + S2\n",
    "            elif self.tr[L][C] == \"E\":\n",
    "                S1 = '-' + S1\n",
    "                S2 = self.s2[C - 1] + S2\n",
    "            else:\n",
    "                S1 = self.s1[L - 1] + S1\n",
    "                S2 = '-' + S2        \n",
    "    \n",
    "            L += DL\n",
    "            C += DC\n",
    "\n",
    "        return S1, S2 \n",
    "    def __repr__(self): \n",
    "        cols = \"-\" + self.s2\n",
    "        lins = \"-\" + self.s1\n",
    "        with io.StringIO(\"\") as S:\n",
    "            print(' ', *cols, sep = '   ', file = S)  \n",
    "            for L, linha in zip(lins, self.mat):\n",
    "                print(L, *[f'{x:3d}' for x in linha], file = S)\n",
    "            print(file = S)\n",
    "\n",
    "            print(' ', *cols, file = S) \n",
    "            for L, linha in zip(lins, self.tr):\n",
    "                print(L, *linha, file = S)\n",
    "\n",
    "            return S.getvalue()\n",
    "        \n",
    "alin= NW(s1_,s2_,-4,3,-1)\n",
    "print (alin)\n",
    "print()\n",
    "print(*alin.rebuild(), sep=\"\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3e94663",
   "metadata": {},
   "source": [
    "### SW:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "287b1e04",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "escolha o valor de gap: -4\n",
      "escolha o valor de match: 3\n",
      "escolha o valor de mismatch: -1\n",
      "    -   G   G   G   A   T   G   G   A   A   G   A   T   G   G   C   G   G   C   C   T   A\n",
      "-   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0\n",
      "C   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   3   0   0   3   3   0   0\n",
      "A   0   0   0   0   3   0   0   0   3   3   0   3   0   0   0   0   2   0   0   2   2   3\n",
      "A   0   0   0   0   3   2   0   0   3   6   2   3   2   0   0   0   0   1   0   0   1   5\n",
      "C   0   0   0   0   0   2   1   0   0   2   5   1   2   1   0   3   0   0   4   3   0   1\n",
      "A   0   0   0   0   3   0   1   0   3   3   1   8   4   1   0   0   2   0   0   3   2   3\n",
      "C   0   0   0   0   0   2   0   0   0   2   2   4   7   3   0   3   0   1   3   3   2   1\n",
      "G   0   3   3   3   0   0   5   3   0   0   5   1   3  10   6   2   6   3   0   2   2   1\n",
      "G   0   3   6   6   2   0   3   8   4   0   3   4   0   6  13   9   5   9   5   1   1   1\n",
      "G   0   3   6   9   5   1   3   6   7   3   3   2   3   3   9  12  12   8   8   4   0   0\n",
      "A   0   0   2   5  12   8   4   2   9  10   6   6   2   2   5   8  11  11   7   7   3   3\n",
      "A   0   0   0   1   8  11   7   3   5  12   9   9   5   1   1   4   7  10  10   6   6   6\n",
      "A   0   0   0   0   4   7  10   6   6   8  11  12   8   4   0   0   3   6   9   9   5   9\n",
      "C   0   0   0   0   0   3   6   9   5   5   7  10  11   7   3   3   0   2   9  12   8   5\n",
      "C   0   0   0   0   0   0   2   5   8   4   4   6   9  10   6   6   2   0   5  12  11   7\n",
      "T   0   0   0   0   0   3   0   1   4   7   3   3   9   8   9   5   5   1   1   8  15  11\n",
      "C   0   0   0   0   0   0   2   0   0   3   6   2   5   8   7  12   8   4   4   4  11  14\n",
      "A   0   0   0   0   3   0   0   1   3   3   2   9   5   4   7   8  11   7   3   3   7  14\n",
      "C   0   0   0   0   0   2   0   0   0   2   2   5   8   4   3  10   7  10  10   6   3  10\n",
      "C   0   0   0   0   0   0   1   0   0   0   1   1   4   7   3   6   9   6  13  13   9   6\n",
      "C   0   0   0   0   0   0   0   0   0   0   0   0   0   3   6   6   5   8   9  16  12   8\n",
      "\n",
      "  - G G G A T G G A A G A T G G C G G C C T A\n",
      "- 0 C C C C C C C C C C C C C C C C C C C C C\n",
      "C E 0 0 0 0 0 0 0 0 0 0 0 0 0 0 D 0 0 D D 0 0\n",
      "A E 0 0 0 D 0 0 0 D D 0 D 0 0 0 0 D 0 0 D D D\n",
      "A E 0 0 0 D D 0 0 D D D D D 0 0 0 0 D 0 0 D D\n",
      "C E 0 0 0 0 D D 0 0 D D D D D 0 D 0 0 D D 0 E\n",
      "A E 0 0 0 D 0 D D D D D D C D D 0 D 0 E D D D\n",
      "C E 0 0 0 0 D 0 D 0 D D E D D D D 0 D D D D D\n",
      "G E D D D 0 0 D D 0 0 D D D D D C D D D D D D\n",
      "G E D D D D 0 D D C C D D D D D C D D C C D D\n",
      "G E D D D D D D D D D D D D D D D D D D D D D\n",
      "A E 0 D D D C C D D D C D C D E D D D D D D D\n",
      "A E 0 0 D D D D D D D D D D D D D D D D D D D\n",
      "A E 0 0 0 D D D D D D D D D D D D D D D D D D\n",
      "C E 0 0 0 E D D D D D D D D D D D 0 D D D D E\n",
      "C E 0 0 0 0 0 D D D D D D D D D D D 0 D D D D\n",
      "T E 0 0 0 0 D 0 D D D D D D D D D D D E E D C\n",
      "C E 0 0 0 0 0 D 0 D D D D E D D D C D D D E D\n",
      "A E 0 0 0 D 0 0 D D D D D C D D E D D D D E D\n",
      "C E 0 0 0 0 D 0 0 D D D E D D D D D D D D E E\n",
      "C E 0 0 0 0 0 D 0 0 0 D D D D D D D D D D C E\n",
      "C E 0 0 0 0 0 0 D 0 0 0 D D D D D D D D D D D\n",
      "\n",
      "ACGGGAAACCTCACCC\n",
      "ATGGAAGATGGCGGCC\n"
     ]
    }
   ],
   "source": [
    "class SW:\n",
    "    def __init__(self, s1, s2, gap_penalty, match_score, mismatch_penalty):\n",
    "        self.s1 = s1 \n",
    "        self.s2 = s2 \n",
    "        self.gap_penalty= gap_penalty\n",
    "        self.match_score= match_score\n",
    "        self.mismatch_penalty= mismatch_penalty\n",
    "        self.score = [[0 for j in range(len(s2) + 1)] for i in range(len(s1) + 1)]\n",
    "        self.tr = [[0 for j in range(len(s2) + 1)] for i in range(len(s1) + 1)]\n",
    "       \n",
    "\n",
    "        for L in range(len(s1)): \n",
    "            self.tr[L + 1][0]  = 'E' \n",
    "        for C in range(len(s2)):\n",
    "            self.tr[0][C + 1]  = 'C'\n",
    "                \n",
    "        for i in range(1, len(self.s1) + 1):\n",
    "            for j in range(1, len(self.s2) + 1):\n",
    "                match = self.score[i - 1][j - 1] + (self.match_score if self.s1[i - 1] == self.s2[j - 1] else self.mismatch_penalty)\n",
    "                delete = self.score[i - 1][j] + self.gap_penalty\n",
    "                insert = self.score[i][j - 1] + self.gap_penalty\n",
    "                self.score[i][j] = max(0, match, delete, insert)\n",
    "                possiveis= (match, delete, insert,0)\n",
    "                dirs = 'DEC0'\n",
    "                self.tr[i][j] = dirs[possiveis.index(self.score[i][j])]\n",
    "                \n",
    "                \n",
    "    def rebuild(self):\n",
    "    \n",
    "        L = len(self.s1)\n",
    "        C = len(self.s2)\n",
    "            \n",
    "        max_i, max_j = 0, 0\n",
    "        max_score = 0\n",
    "        for i in range(1, L + 1):\n",
    "            for j in range(1, C + 1):\n",
    "                if self.score[i][j] > max_score:\n",
    "                    max_i, max_j = i, j\n",
    "                    max_score = self.score[i][j]\n",
    "        i, j = max_i, max_j\n",
    "        alignD1= \"\" \n",
    "        alignD2= \"\"\n",
    "        while self.score[i][j] != 0:\n",
    "            current_score = self.score[i][j]\n",
    "            diagonal_score = self.score[i - 1][j - 1]\n",
    "            up_score = self.score[i][j - 1]\n",
    "            left_score = self.score[i - 1][j]\n",
    "\n",
    "            if current_score == diagonal_score + (self.match_score if self.s1[i - 1] == self.s2[j - 1] else self.mismatch_penalty):\n",
    "                alignD1 += self.s1[i - 1]\n",
    "                alignD2 += self.s2[j - 1]\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "            elif current_score == left_score + self.gap_penalty:\n",
    "                alignD1 += self.s1[i - 1]\n",
    "                alignD2 += \"-\"\n",
    "                i -= 1\n",
    "            elif current_score == up_score + self.gap_penalty:\n",
    "                alignD1 += \"-\"\n",
    "                alignD2 += self.s2[j - 1]\n",
    "                j -= 1\n",
    "        \n",
    "                \n",
    "        return alignD1[::-1], alignD2[::-1] \n",
    "    \n",
    "    \n",
    "    \n",
    "    def __repr__(self): \n",
    "        cols = \"-\" + self.s2\n",
    "        lins = \"-\" + self.s1\n",
    "        with io.StringIO(\"\") as S:\n",
    "            print(' ', *cols, sep = '   ', file = S)  \n",
    "            for L, linha in zip(lins, self.score):\n",
    "                print(L, *[f'{x:3d}' for x in linha], file = S)\n",
    "            print(file = S)\n",
    "\n",
    "            print(' ', *cols, file = S) \n",
    "            for L, linha in zip(lins, self.tr ):\n",
    "                print(L, *linha, file = S)\n",
    "\n",
    "            return S.getvalue()\n",
    "\n",
    "es_gap= int(input('escolha o valor de gap: '))\n",
    "es_match= int(input('escolha o valor de match: '))\n",
    "es_mis=int(input('escolha o valor de mismatch: '))\n",
    "\n",
    "alignSW= SW(s1_,s2_,es_gap,es_match,es_mis)\n",
    "print(alignSW)\n",
    "print(*alignSW.rebuild(), sep=\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "76a97c9c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    -   T   A   T   G   C\n",
      "-   0  -4  -8 -12 -16 -20\n",
      "A  -4   0  -4  -4  -8 -12\n",
      "A  -8  -4   0   0  -4  -8\n",
      "T -12  -8  -4   4   0  -4\n",
      "G -16 -12  -3   0   9   5\n",
      "A -20 -16  -7  -3   5  15\n",
      "\n",
      "  - T A T G C\n",
      "- 0 E E E E E\n",
      "A C D D D E E\n",
      "A C D D D D D\n",
      "T C D D D D D\n",
      "G C C D C D E\n",
      "A C C C D C D\n",
      "\n",
      "\n",
      "AATGA\n",
      "TATGC\n"
     ]
    }
   ],
   "source": [
    "#Teste Blossum:\n",
    "\n",
    "# Lê a matriz blosum 62 de um ficheiro na diretoria\n",
    "\n",
    "def create_blossum62():\n",
    "    mat = \"\"\"\n",
    "       A  R  N  D  C  Q  E  G  H  I  L  K  M  F  P  S  T  W  Y  V  B  Z  X  *\n",
    "    A  4 -1 -2 -2  0 -1 -1  0 -2 -1 -1 -1 -1 -2 -1  1  0 -3 -2  0 -2 -1  0 -4 \n",
    "    R -1  5  0 -2 -3  1  0 -2  0 -3 -2  2 -1 -3 -2 -1 -1 -3 -2 -3 -1  0 -1 -4 \n",
    "    N -2  0  6  1 -3  0  0  0  1 -3 -3  0 -2 -3 -2  1  0 -4 -2 -3  3  0 -1 -4 \n",
    "    D -2 -2  1  6 -3  0  2 -1 -1 -3 -4 -1 -3 -3 -1  0 -1 -4 -3 -3  4  1 -1 -4 \n",
    "    C  0 -3 -3 -3  9 -3 -4 -3 -3 -1 -1 -3 -1 -2 -3 -1 -1 -2 -2 -1 -3 -3 -2 -4 \n",
    "    Q -1  1  0  0 -3  5  2 -2  0 -3 -2  1  0 -3 -1  0 -1 -2 -1 -2  0  3 -1 -4 \n",
    "    E -1  0  0  2 -4  2  5 -2  0 -3 -3  1 -2 -3 -1  0 -1 -3 -2 -2  1  4 -1 -4 \n",
    "    G  0 -2  0 -1 -3 -2 -2  6 -2 -4 -4 -2 -3 -3 -2  0 -2 -2 -3 -3 -1 -2 -1 -4 \n",
    "    H -2  0  1 -1 -3  0  0 -2  8 -3 -3 -1 -2 -1 -2 -1 -2 -2  2 -3  0  0 -1 -4 \n",
    "    I -1 -3 -3 -3 -1 -3 -3 -4 -3  4  2 -3  1  0 -3 -2 -1 -3 -1  3 -3 -3 -1 -4 \n",
    "    L -1 -2 -3 -4 -1 -2 -3 -4 -3  2  4 -2  2  0 -3 -2 -1 -2 -1  1 -4 -3 -1 -4 \n",
    "    K -1  2  0 -1 -3  1  1 -2 -1 -3 -2  5 -1 -3 -1  0 -1 -3 -2 -2  0  1 -1 -4 \n",
    "    M -1 -1 -2 -3 -1  0 -2 -3 -2  1  2 -1  5  0 -2 -1 -1 -1 -1  1 -3 -1 -1 -4 \n",
    "    F -2 -3 -3 -3 -2 -3 -3 -3 -1  0  0 -3  0  6 -4 -2 -2  1  3 -1 -3 -3 -1 -4 \n",
    "    P -1 -2 -2 -1 -3 -1 -1 -2 -2 -3 -3 -1 -2 -4  7 -1 -1 -4 -3 -2 -2 -1 -2 -4 \n",
    "    S  1 -1  1  0 -1  0  0  0 -1 -2 -2  0 -1 -2 -1  4  1 -3 -2 -2  0  0  0 -4 \n",
    "    T  0 -1  0 -1 -1 -1 -1 -2 -2 -1 -1 -1 -1 -2 -1  1  5 -2 -2  0 -1 -1  0 -4 \n",
    "    W -3 -3 -4 -4 -2 -2 -3 -2 -2 -3 -2 -3 -1  1 -4 -3 -2 11  2 -3 -4 -3 -2 -4 \n",
    "    Y -2 -2 -2 -3 -2 -1 -2 -3  2 -1 -1 -2 -1  3 -3 -2 -2  2  7 -1 -3 -2 -1 -4 \n",
    "    V  0 -3 -3 -3 -1 -2 -2 -3 -3  3  1 -2  1 -1 -2 -2  0 -3 -1  4 -3 -2 -1 -4 \n",
    "    B -2 -1  3  4 -3  0  1 -1  0 -3 -4  0 -3 -3 -2  0 -1 -4 -3 -3  4  1 -1 -4 \n",
    "    Z -1  0  0  1 -3  3  4 -2  0 -3 -3  1 -1 -3 -1  0 -1 -3 -2 -2  1  4 -1 -4 \n",
    "    X  0 -1 -1 -1 -2 -1 -1 -1 -1 -1 -1 -1 -1 -1 -2  0  0 -2 -1 -1 -1 -1 -1 -4 \n",
    "    * -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4  1\"\"\"\n",
    "\n",
    "    D = {}\n",
    "\n",
    "    fields, *M = [L.strip().split() for L in mat.splitlines() if L.strip() != '']\n",
    "\n",
    "    for L in M:\n",
    "        k, *L = L\n",
    "        for f, v in zip(fields, L):\n",
    "            if f not in D:\n",
    "                D[f] = {}\n",
    "            D[f][k] = int(v)\n",
    "    return D\n",
    "\n",
    "sm = create_blossum62()\n",
    "\n",
    "# score of a position (column) \n",
    "def score_pos (c1, c2, sm, gap_penalty):\n",
    "    if c1 == \"-\" or c2==\"-\":\n",
    "        return gap_penalty\n",
    "    else:\n",
    "        return sm[c1][c2]\n",
    "    \n",
    "def score_align (seq1, seq2, sm, g):\n",
    "    res = 0;\n",
    "    for i in range(len(seq1)):\n",
    "        res += score_pos (seq1[i], seq2[i], sm, g)\n",
    "    return res\n",
    "\n",
    "\n",
    "class NW_blossum:\n",
    "    def __init__(self, s1, s2, sm, gap_penalty):\n",
    "        self.s1 = s1 \n",
    "        self.s2 = s2 \n",
    "        self.sm= sm\n",
    "        self.gap_penalty= gap_penalty\n",
    "        self.mat = Mat(len(s1) + 1, len(s2) + 1)\n",
    "        self.tr  = Mat(len(s1) + 1, len(s2) + 1)\n",
    "\n",
    "        for L in range(len(s1)):\n",
    "            self.mat[L + 1][0] = self.gap_penalty * (L + 1) \n",
    "            self.tr[L + 1][0]  = 'C' \n",
    "        for C in range(len(s2)):\n",
    "            self.mat[0][C + 1] = self.gap_penalty * (C + 1) \n",
    "            self.tr[0][C + 1]  = 'E'\n",
    "\n",
    "        for L, x1 in enumerate(s1):\n",
    "            for C, x2 in enumerate(s2):\n",
    "                possiveis = [\n",
    "                    self.mat[L  ][C    ] + score_pos (self.s1[L-1], self.s2[C-1], self.sm, self.gap_penalty),   # Diagonal\n",
    "                    self.mat[L+1][C    ] + self.gap_penalty,               # Esquerda\n",
    "                    self.mat[L  ][C + 1] + self.gap_penalty,               # Cima\n",
    "                ]\n",
    "                dirs = \"DEC\"\n",
    "                self.mat[L + 1][C + 1] = max(possiveis)\n",
    "                self.tr[L + 1][C + 1] = dirs[possiveis.index(self.mat[L + 1][C + 1])]\n",
    "                \n",
    "    def rebuild(self):\n",
    "        L = len(self.s1)\n",
    "        C = len(self.s2)\n",
    "        S1 = \"\"\n",
    "        S2 = \"\"\n",
    "    \n",
    "        dirs = {\n",
    "            'D' : (-1, -1),\n",
    "            'E' : ( 0, -1),\n",
    "            'C' : (-1,  0)\n",
    "        }\n",
    "        while L > 0 or C > 0:\n",
    "            \n",
    "            DL, DC = dirs[self.tr[L][C]]\n",
    "            if self.tr[L][C] == \"D\":\n",
    "                S1 = self.s1[L - 1] + S1 \n",
    "                S2 = self.s2[C - 1] + S2\n",
    "            elif self.tr[L][C] == \"E\":\n",
    "                S1 = '-' + S1\n",
    "                S2 = self.s2[C - 1] + S2\n",
    "            else:\n",
    "                S1 = self.s1[L - 1] + S1\n",
    "                S2 = '-' + S2        \n",
    "    \n",
    "            L += DL\n",
    "            C += DC\n",
    "\n",
    "        return S1, S2 \n",
    "    def __repr__(self): \n",
    "        cols = \"-\" + self.s2\n",
    "        lins = \"-\" + self.s1\n",
    "        with io.StringIO(\"\") as S:\n",
    "            print(' ', *cols, sep = '   ', file = S)  \n",
    "            for L, linha in zip(lins, self.mat):\n",
    "                print(L, *[f'{x:3d}' for x in linha], file = S)\n",
    "            print(file = S)\n",
    "\n",
    "            print(' ', *cols, file = S) \n",
    "            for L, linha in zip(lins, self.tr):\n",
    "                print(L, *linha, file = S)\n",
    "\n",
    "            return S.getvalue()\n",
    "\n",
    "        \n",
    "Seq= 'AATGA'\n",
    "Seq1= 'TATGC'\n",
    "alin= NW(Seq,Seq1,sm,-4)\n",
    "print (alin)\n",
    "print()\n",
    "print(*alin.rebuild(), sep=\"\\n\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "a997c2fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    -   T   A   T   G   C\n",
      "-   0   0   0   0   0   0\n",
      "A   0   0   4   0   0   0\n",
      "A   0   0   4   4   0   0\n",
      "T   0   5   1   9   5   1\n",
      "G   0   1   5   5  15  11\n",
      "A   0   0   5   5  11  15\n",
      "\n",
      "  - T A T G C\n",
      "- 0 C C C C C\n",
      "A E D D D D D\n",
      "A E D D D D D\n",
      "T E D C D C C\n",
      "G E E D E D C\n",
      "A E D D D E D\n",
      "\n",
      "ATG\n",
      "ATG\n"
     ]
    }
   ],
   "source": [
    "class SW_blossum:\n",
    "    def __init__(self, s1, s2, sm, gap_penalty):\n",
    "        self.s1 = s1 \n",
    "        self.s2 = s2 \n",
    "        self.sm= sm\n",
    "        self.gap_penalty= gap_penalty\n",
    "        self.score = [[0 for j in range(len(s2) + 1)] for i in range(len(s1) + 1)]\n",
    "        self.tr = [[0 for j in range(len(s2) + 1)] for i in range(len(s1) + 1)]\n",
    "       \n",
    "\n",
    "        for L in range(len(s1)): \n",
    "            self.tr[L + 1][0]  = 'E' \n",
    "        for C in range(len(s2)):\n",
    "            self.tr[0][C + 1]  = 'C'\n",
    "                \n",
    "        for i in range(1, len(self.s1) + 1):\n",
    "            for j in range(1, len(self.s2) + 1):\n",
    "                match = self.score[i - 1][j - 1] + score_pos (self.s1[i-1], self.s2[j-1], self.sm, self.gap_penalty)\n",
    "                delete = self.score[i - 1][j] + self.gap_penalty\n",
    "                insert = self.score[i][j - 1] + self.gap_penalty\n",
    "                self.score[i][j] = max(0, match, delete, insert)\n",
    "                possiveis= (match, delete, insert,0)\n",
    "                dirs = 'DEC0'\n",
    "                self.tr[i][j] = dirs[possiveis.index(self.score[i][j])]\n",
    "                \n",
    "                \n",
    "    def rebuild(self):\n",
    "    \n",
    "        L = len(self.s1)\n",
    "        C = len(self.s2)\n",
    "            \n",
    "        max_i, max_j = 0, 0\n",
    "        max_score = 0\n",
    "        for i in range(1, L + 1):\n",
    "            for j in range(1, C + 1):\n",
    "                if self.score[i][j] > max_score:\n",
    "                    max_i, max_j = i, j\n",
    "                    max_score = self.score[i][j]\n",
    "        i, j = max_i, max_j\n",
    "        alignD1= \"\" \n",
    "        alignD2= \"\"\n",
    "        while self.score[i][j] != 0:\n",
    "            current_score = self.score[i][j]\n",
    "            diagonal_score = self.score[i - 1][j - 1]\n",
    "            up_score = self.score[i][j - 1]\n",
    "            left_score = self.score[i - 1][j]\n",
    "\n",
    "            if current_score == diagonal_score + score_pos (self.s1[i-1], self.s2[j-1], self.sm, self.gap_penalty):\n",
    "                alignD1 += self.s1[i - 1]\n",
    "                alignD2 += self.s2[j - 1]\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "            elif current_score == left_score + self.gap_penalty:\n",
    "                alignD1 += self.s1[i - 1]\n",
    "                alignD2 += \"-\"\n",
    "                i -= 1\n",
    "            elif current_score == up_score + self.gap_penalty:\n",
    "                alignD1 += \"-\"\n",
    "                alignD2 += self.s2[j - 1]\n",
    "                j -= 1\n",
    "        \n",
    "                \n",
    "        return alignD1[::-1], alignD2[::-1] \n",
    "    \n",
    "    \n",
    "    \n",
    "    def __repr__(self): \n",
    "        cols = \"-\" + self.s2\n",
    "        lins = \"-\" + self.s1\n",
    "        with io.StringIO(\"\") as S:\n",
    "            print(' ', *cols, sep = '   ', file = S)  \n",
    "            for L, linha in zip(lins, self.score):\n",
    "                print(L, *[f'{x:3d}' for x in linha], file = S)\n",
    "            print(file = S)\n",
    "\n",
    "            print(' ', *cols, file = S) \n",
    "            for L, linha in zip(lins, self.tr ):\n",
    "                print(L, *linha, file = S)\n",
    "\n",
    "            return S.getvalue()\n",
    "\n",
    "#es_gap= int(input('escolha o valor de gap: '))\n",
    "#es_match= int(input('escolha o valor de match: '))\n",
    "#es_mis=int(input('escolha o valor de mismatch: '))\n",
    "\n",
    "Seq= 'AATGA'\n",
    "Seq1= 'TATGC'\n",
    "alignSW= SW_blossum(Seq,Seq1,sm,-4)\n",
    "print(alignSW)\n",
    "print(*alignSW.rebuild(), sep=\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "d8a9ff68",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Prefere utilizar blossum ou definir o input(Definir- você define o input, Blossum- utiliza blossum): Blossum\n",
      "escolha o valor de gap: -4\n",
      "    -   G   G   G   A   T   G   G   A   A   G   A   T   G   G   C   G   G   C   C   T   A\n",
      "-   0  -4  -8 -12 -16 -20 -24 -28 -32 -36 -40 -44 -48 -52 -56 -60 -64 -68 -72 -76 -80 -84\n",
      "C  -4   0  -4  -8 -12 -16 -20 -24 -28 -32 -36 -40 -44 -48 -52 -56 -51 -55 -59 -63 -67 -71\n",
      "A  -8  -4  -3  -7 -11 -12 -16 -20 -24 -28 -32 -36 -40 -44 -48 -52 -47 -51 -55 -50 -54 -58\n",
      "A -12  -4  -4  -3  -7  -7 -11 -15 -19 -20 -24 -28 -32 -36 -40 -44 -48 -47 -51 -54 -50 -54\n",
      "C -16  -8  -4  -4  -3  -3  -7 -11 -15 -15 -16 -20 -24 -28 -32 -36 -40 -44 -47 -51 -54 -50\n",
      "A -20 -12  -8  -7  -7  -3  -4  -8 -12 -15 -15 -19 -20 -24 -28 -32 -27 -31 -35 -38 -42 -46\n",
      "C -24 -16 -12  -8  -7  -3  -3  -4  -8  -8 -11 -15 -15 -19 -23 -27 -31 -27 -31 -35 -38 -42\n",
      "G -28 -20 -16 -12 -11  -7  -4  -6  -7  -8  -8 -12 -15 -16 -20 -24 -18 -22 -26 -22 -26 -30\n",
      "G -32 -24 -14 -10  -6 -10  -8   2   0  -4  -8  -2  -6 -10 -10 -14 -18 -12 -16 -20 -24 -28\n",
      "G -36 -28 -18  -8  -4  -6 -10  -2   8   4   0  -2  -2  -6  -4  -4  -8 -12  -6 -10 -14 -18\n",
      "A -40 -32 -22 -12  -2  -4  -8  -4   4   8   4   6   2  -2   0   2  -2  -2  -6  -9 -13 -16\n",
      "A -44 -36 -26 -16  -6   2  -2  -6   0   8  12   8  10   6   2   0   2  -2  -2  -6  -9 -13\n",
      "A -48 -40 -30 -20 -10  -2   2  -2  -4   4  12  12  12  10   6   2   0   2  -2  -2  -6  -9\n",
      "C -52 -44 -34 -24 -14  -6  -2   2  -2   0   8  12  16  12  10   6   2   0   2  -2  -2  -6\n",
      "C -56 -48 -38 -28 -18 -10  -6  -2  -1  -2   4   8  12  15  11   7  15  11   7  11   7   3\n",
      "T -60 -52 -42 -32 -22 -14 -10  -6  -5  -1   0   4   8  11  12   8  16  12   8  16  20  16\n",
      "C -64 -56 -46 -36 -26 -18  -9 -10  -8  -5  -1   0   4  13   9  10  12  14  10  12  16  25\n",
      "A -68 -60 -50 -40 -30 -22 -13 -12 -12  -8  -5  -4   0   9  10   6  19  15  11  19  21  21\n",
      "C -72 -64 -54 -44 -34 -26 -17 -13 -12  -8  -4  -5   0   5   9  10  15  19  15  15  19  21\n",
      "C -76 -68 -58 -48 -38 -30 -21 -17 -16 -12  -8  -7  -4   1   5   6  19  15  16  24  24  20\n",
      "C -80 -72 -62 -52 -42 -34 -25 -21 -20 -16 -12 -11  -7  -3   1   2  15  16  12  25  33  29\n",
      "\n",
      "  - G G G A T G G A A G A T G G C G G C C T A\n",
      "- 0 E E E E E E E E E E E E E E E E E E E E E\n",
      "C C D E E E D E E E D D E D E E E D E E D D E\n",
      "A C D D D D D E E E D D E D E E E D E E D D E\n",
      "A C D D D D D E E E D D E D E E E E D D C D D\n",
      "C C D D D D D D D D D D E D E E E E E D D D D\n",
      "A C C C D D D D E E D D D D E E E D E E D D E\n",
      "C C D D D D D D D D D D D D E E E E D D D D D\n",
      "G C C C C D D D D D D D E D D E E D E E D D E\n",
      "G C C D D D E C D D E D D E E D D E D D E E D\n",
      "G C C D D D D E D D E E D D E D D E D D E E E\n",
      "A C C D D D D D D D D D D E E D D E D D D D D\n",
      "A C D C C C D E E C D D E D E E D D D D D D D\n",
      "A C D C C C D D D C D D D D D D D D D D D D D\n",
      "C C D C C C D D D D D D D D D D D D D D D D D\n",
      "C C C C C C C C C D D C C D D E D D E E D D E\n",
      "T C C C C C C C C D D C C D D D D D D D D D E\n",
      "C C C C C C C D C D D D C D D D D C D D C C D\n",
      "A C C C C C C C D C D D D D C D D D E D D D C\n",
      "C C D C C C D C D D D D D D C D D C D D C D D\n",
      "C C C C C C C C C D D D D C C C D D E D D D E\n",
      "C C C C C C C C C D D D D D C C D D D D D D E\n",
      "\n",
      "CAACACGGGAAACCTCA-CCC-\n",
      "-GGGATGGAAGATGGCGGCCTA\n"
     ]
    }
   ],
   "source": [
    "escolha_de_input= input('Prefere utilizar blossum ou definir o input(Definir- você define o input, Blossum- utiliza blossum): ')\n",
    "\n",
    "def_map = {\"Definir\": NW, \"Blossum\": NW_blossum}\n",
    "try:\n",
    "    escolja_cls = def_map[escolha_de_input]\n",
    "    if escolja_cls == NW:\n",
    "        es_gap= int(input('escolha o valor de gap: '))\n",
    "        es_match= int(input('escolha o valor de match: '))\n",
    "        es_mis=int(input('escolha o valor de mismatch: '))\n",
    "        alin= escolja_cls(s1_,s2_, es_gap ,es_match ,es_mis)\n",
    "        print(alin)\n",
    "        print()\n",
    "        print(*alin.rebuild(), sep=\"\\n\")\n",
    "        \n",
    "    else:\n",
    "        es_gap= int(input('escolha o valor de gap: '))\n",
    "        align= NW_blossum(s1_,s2_,sm,-4)\n",
    "        print(align)\n",
    "        print(*align.rebuild(), sep=\"\\n\")\n",
    "except KeyError:\n",
    "    raise ValueError(\"Tipo de alinhamento inválido.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84ce5d86",
   "metadata": {},
   "source": [
    "### Multiple align:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "967e6737",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "escolha o tipo de alinhamento que quer(SW ou NW): SW\n",
      "escolha o valor de gap: -4\n",
      "escolha o valor de match: 3\n",
      "escolha o valor de mismatch: -1\n",
      "\n",
      "[\"('AAADK0043071.1', 21)\", \"('AAADS0003348.1', 20)\", \"('AAADW0009410.1', 21)\", \"('AAADX0043451.1', 21)\", \"('ABAAM0346030.1', 20)\"]\n",
      "escolha a seq da BD que pretende adicionar ao alinhamento: 4\n",
      "\n",
      "ACGGGAAACCTCACCC\n",
      "ATGGAAGATGGCGGCC\n",
      "\n",
      "ACGGGAAACCTCACCC\n",
      "\n",
      "GGGAAACCTCACCC\n",
      "GGTGAACTTCTGCC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def consensus(s1_, s2_):\n",
    "    res = \"\"\n",
    "    for x, y in zip(s1_, s2_):\n",
    "        if x == y:\n",
    "            res += x\n",
    "        elif x == '-':\n",
    "            res += y\n",
    "        else:\n",
    "            res += x\n",
    "    return res\n",
    "\n",
    "escolha_alinh= input('escolha o tipo de alinhamento que quer(SW ou NW): ')\n",
    "al_map = {\"SW\": SW, \"NW\": NW}\n",
    "\n",
    "try:\n",
    "    alinhamento_cls = alg_map[escolha_alinh]\n",
    "except KeyError:\n",
    "    raise ValueError(\"Tipo de alinhamento inválido.\")\n",
    "                     \n",
    "es_gap= int(input('escolha o valor de gap: '))\n",
    "es_match= int(input('escolha o valor de match: '))\n",
    "es_mis=int(input('escolha o valor de mismatch: '))\n",
    "print()\n",
    "print (Seq_and_length)\n",
    "seq3= int(input('escolha a seq da BD que pretende adicionar ao alinhamento: '))\n",
    "print()\n",
    "s3= todas_as_seq[seq3]            \n",
    "a1, a2 = alinhamento_cls(s1_,s2_,es_gap,es_match,es_mis).rebuild()\n",
    "print(a1)\n",
    "print(a2)\n",
    "print()\n",
    "m1 = consensus(a1,a2)\n",
    "print(m1)\n",
    "print()\n",
    "m2, a3 = alinhamento_cls(m1, s3, es_gap,es_match,es_mis).rebuild()\n",
    "print(m2,a3, sep=\"\\n\")\n",
    "print()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "360c59cf",
   "metadata": {},
   "source": [
    "## Enzimas de corte:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "6ec9c928",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'R': '[GA]', 'Y': '[CT]', 'M': '[AC]', 'K': '[GT]', 'S': '[GC]', 'W': '[AT]', 'B': '[^A]', 'D': '[^C]', 'H': '[^G]', 'V': '[^T]', 'N': '[ACGT]'}\n",
      "\n",
      "selecione o que deseja cortar: R^K\n",
      "\n",
      "[\"('AAADK0043071.1', 21)\", \"('AAADS0003348.1', 20)\", \"('AAADW0009410.1', 21)\", \"('AAADX0043451.1', 21)\", \"('ABAAM0346030.1', 20)\"]\n",
      "escolha o id da BD que pretende analisar: 4\n",
      "['G', 'TTTG', 'G', 'TGAACTTCTGCCTA']\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "enzimas = {\"R\" : \"[GA]\", \"Y\" : \"[CT]\", \"M\" : \"[AC]\", \"K\" : \"[GT]\", \"S\" : \"[GC]\", \"W\" : \"[AT]\", \"B\" : \"[^A]\", \"D\" : \"[^C]\", \"H\" : \"[^G]\", \"V\" : \"[^T]\", \"N\" : \"[ACGT]\"}\n",
    "\n",
    "print (enzimas)\n",
    "print()\n",
    "enzima = input('selecione o que deseja cortar: ')\n",
    "print()\n",
    "\n",
    "print (Seq_and_length)\n",
    "seq_enzimas= int(input('escolha o id da BD que pretende analisar: '))\n",
    "s_enzimas= todas_as_seq[seq_enzimas] \n",
    "\n",
    "def converte(enz):\n",
    "    return ''.join([enzimas.get(x, x) for x in enz])\n",
    "\n",
    "def enz_corte(enzima):\n",
    "    e1, e2 = enzima.split('^')\n",
    "    return f\"(?<={converte(e1)}){converte(e2)}\"\n",
    "\n",
    "pos = [m.span()[0] for m in re.finditer(enz_corte(enzima), str(s_enzimas))]\n",
    "\n",
    "a=[str(s_enzimas[i:j]) for i, j in zip([0, *pos], [*pos, len(str(s_enzimas))])]\n",
    "print()\n",
    "print(a)\n",
    "    \n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "4b059626",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GGGATGGAAGATGGCGGCCTA\n"
     ]
    }
   ],
   "source": [
    "print (str(s1_))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fa0d69e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "class Prosite:\n",
    "    def __init__(self, input_enzyme, query):\n",
    "        self.enzymes = {\"R\" : \"[GA]\", \"Y\" : \"[CT]\", \"M\" : \"[AC]\", \"K\" : \"[GT]\", \"S\" : \"[GC]\", \"W\" : \"[AT]\", \"B\" : \"[^A]\", \"D\" : \"[^C]\", \"H\" : \"[^G]\", \"V\" : \"[^T]\", \"N\" : \"[ACGT]\"}\n",
    "        self.input_enzyme = input_enzyme\n",
    "        self.query = query\n",
    "        \n",
    "    def valid_input(self):\n",
    "        count = 0\n",
    "        for i in self.input_enzyme:\n",
    "            if i == '^':\n",
    "                count += 1\n",
    "        if count == 0:\n",
    "            raise Exception(\"The input enzyme is invalid. It must contain 1 '^' character.\")\n",
    "            return Exception\n",
    "        elif count >= 2:\n",
    "            raise Exception(\"The input enzyme is invalid. It must contain 1 '^' character.\")\n",
    "            return Exception\n",
    "        \n",
    "    def convert(self, part):\n",
    "        return ''.join([self.enzymes.get(x, x) for x in part])\n",
    "    \n",
    "    def enz_cut(self):\n",
    "        if self.valid_input() != \"Exception\":\n",
    "            e1, e2 = self.input_enzyme.split('^')\n",
    "            return f\"(?<={self.convert(e1)}){self.convert(e2)}\"\n",
    "        else:\n",
    "            return self.valid_input()\n",
    "        \n",
    "    def __str__(self):\n",
    "        pos = [m.span()[0] for m in re.finditer(self.enz_cut(), str(self.query))]\n",
    "        a = [str(self.query[i:j]) for i, j in zip([0, *pos], [*pos, len(str(self.query))])]\n",
    "        return str(a)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "5d95abc6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "test_enz_cut (__main__.TestProsite) ... ok\n",
      "test_repr (__main__.TestProsite) ... ok\n",
      "test_valid_input (__main__.TestProsite) ... ok\n",
      "\n",
      "----------------------------------------------------------------------\n",
      "Ran 3 tests in 0.005s\n",
      "\n",
      "OK\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<unittest.runner.TextTestResult run=3 errors=0 failures=0>"
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import unittest\n",
    "class TestProsite(unittest.TestCase):\n",
    "    def test_valid_input(self):\n",
    "        with self.assertRaises(Exception) as error:\n",
    "            self.tmp1= Prosite(\"RK\",\"GTTTGGTGAACTTCTGCCTA\").valid_input()\n",
    "            self.tmp2= Prosite(\"R^^K\",\"GTTTGGTGAACTTCTGCCTA\").valid_input()\n",
    "        self.assertEqual(error.exception.args[0], \"The input enzyme is invalid. It must contain 1 '^' character.\")\n",
    "        \n",
    "    def test_enz_cut(self):\n",
    "        tmp = Prosite(\"R^K\",\"GTTTGGTGAACTTCTGCCTA\")\n",
    "        self.assertEqual(tmp.enz_cut(),'(?<=[GA])[GT]')\n",
    "        with self.assertRaises(Exception) as error:\n",
    "            self.tmp= Prosite(\"RK\",\"GTTTGGTGAACTTCTGCCTA\").enz_cut()\n",
    "        self.assertEqual(error.exception.args[0], \"The input enzyme is invalid. It must contain 1 '^' character.\")\n",
    "        \n",
    "    def test_repr(self):\n",
    "        tmp = Prosite(\"R^K\",\"GTTTGGTGAACTTCTGCCTA\")\n",
    "        self.assertEqual(str(tmp), \"['G', 'TTTG', 'G', 'TGAACTTCTGCCTA']\")\n",
    "        self.assertIsNotNone(str(tmp))\n",
    "        self.assertGreater(len(str(tmp)), 0)\n",
    "        \n",
    "suite = unittest.TestLoader().loadTestsFromTestCase(TestProsite)\n",
    "unittest.TextTestRunner( verbosity=3 ).run( suite )  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "652aa343",
   "metadata": {},
   "outputs": [],
   "source": [
    "enzimas = {\"R\" : \"[GA]\", \"Y\" : \"[CT]\", \"M\" : \"[AC]\", \"K\" : \"[GT]\", \"S\" : \"[GC]\", \"W\" : \"[AT]\", \"B\" : \"[^A]\", \"D\" : \"[^C]\", \"H\" : \"[^G]\", \"V\" : \"[^T]\", \"N\" : \"[ACGT]\"}\n",
    "\n",
    "print (enzimas)\n",
    "print()\n",
    "enzima = input('Insert your cut enzyme (example: R^K): ')\n",
    "print()\n",
    "\n",
    "print(result)\n",
    "seq_enzimas= int(input('escolha o id da BD que pretende analisar: '))\n",
    "s_enzimas= todas_as_seq[seq_enzimas] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "9d3102ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The results from the 2 - Local and Global Alignments:\n",
      "[[0, 'A00829.1', 50], [1, 'AAADK0043071.1', 21], [2, 'AAADS0003348.1', 20], [3, 'AAADW0009410.1', 21], [4, 'AAADX0043451.1', 21], [5, 'AAVX01203288.1', 50], [6, 'ABAAM0346030.1', 20], [7, 'BAAF03073813.1', 50], [8, 'DH405380.1', 50], [9, 'DS196524.1', 50], [10, 'EC923692.1', 50], [11, 'EC966285.1', 50], [12, 'M97702.1', 50], [13, 'X65482.1', 50], [14, 'X68316.1', 50]]\n",
      "\n",
      "Dictionary for the cut enzyme: {'R': '[GA]', 'Y': '[CT]', 'M': '[AC]', 'K': '[GT]', 'S': '[GC]', 'W': '[AT]', 'B': '[^A]', 'D': '[^C]', 'H': '[^G]', 'V': '[^T]', 'N': '[ACGT]'}\n",
      "Insert your cut enzyme (example: R^K): r^k\n",
      "\n",
      "Choose the identificator of a seq-1 that you want to work with: 0\n",
      "\n",
      "TACCTTGTCCGAAAAGGAAAGACAAATCAAGAAGCAAACTGCTTTGGTTG\n",
      "['TACCTTG', 'TCCGAAAA', 'G', 'GAAA', 'GACAAA', 'TCAA', 'GAA', 'GCAAACTGCTTTG', 'G', 'TTG']\n"
     ]
    }
   ],
   "source": [
    "# input_enzyme = input('Insert your cut enzyme (example: R^K): ')\n",
    "print('\\nThe results from the 2 - Local and Global Alignments:\\n{}\\n'.format(result))\n",
    "print(\"Dictionary for the cut enzyme:\", {\"R\" : \"[GA]\", \"Y\" : \"[CT]\", \"M\" : \"[AC]\", \"K\" : \"[GT]\", \"S\" : \"[GC]\", \"W\" : \"[AT]\", \"B\" : \"[^A]\", \"D\" : \"[^C]\", \"H\" : \"[^G]\", \"V\" : \"[^T]\", \"N\" : \"[ACGT]\"})\n",
    "\n",
    "input_enzyme = input('Insert your cut enzyme (example: R^K): ')\n",
    "\n",
    "while True:    \n",
    "    choosep = input('\\nChoose the identificator of a seq-1 that you want to work with: ')\n",
    "    if not choosep.isdigit():\n",
    "        print(\"Error: Choose a valid identificator\")\n",
    "        continue\n",
    "    choosep = int(choosep)\n",
    "    identificators = [i[0] for i in result]\n",
    "    if choosep not in identificators:\n",
    "        print(\"Error: Choose an identificator present in the list\")\n",
    "        continue\n",
    "    else:\n",
    "        s1 = ''.join(db.extract_sequences()[int(choosep)])\n",
    "        print('\\n{}'.format(s1))\n",
    "        print(Prosite(input_enzyme.upper(), s1))\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "484bd77c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GTTTGGTGAACTTCTGCCTA']\n"
     ]
    }
   ],
   "source": [
    "a=Prosite(\"X^A\",\"GTTTGGTGAACTTCTGCCTA\")\n",
    "print(str(a))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
