{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "01927ff8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIXML \n",
    "from Bio.Blast import NCBIWWW\n",
    "import requests\n",
    "import re\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "from Bio import Medline\n",
    "from tqdm import tqdm\n",
    "from Bio import SearchIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "from bs4 import BeautifulSoup\n",
    "from Bio.SeqFeature import CompoundLocation\n",
    "import mysql.connector as SQLC\n",
    "import mysql.connector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "e4a6a6ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Escolha o tamanho máximo das seq: 50\n",
      "[\"('AAADK0043071.1', 21)\", \"('AAADS0003348.1', 20)\", \"('AAADW0009410.1', 21)\", \"('AAADX0043451.1', 21)\", \"('ABAAM0346030.1', 20)\"]\n"
     ]
    }
   ],
   "source": [
    "DataBase = SQLC.connect(\n",
    "    host =\"geo.di.uminho.pt\",\n",
    "    user =\"bioinformatica\",\n",
    "    password =\"20221207\",\n",
    "    database =\"AP_db_KRG\"\n",
    ")\n",
    "DataBase.autocommit = True \n",
    "Cursor = DataBase.cursor()\n",
    "\n",
    "num= input(\"Escolha o tamanho máximo das seq: \")\n",
    "Seq_and_length=[]\n",
    "just_seq=[]\n",
    "try:\n",
    "    sql_Seq_length= f\"select Gene.ID_genebank, Gene.length from Gene where Gene.length <{num}\"\n",
    "    Cursor.execute(sql_Seq_length)\n",
    "    for row in Cursor:\n",
    "        Seq_and_length.append(str(row)) \n",
    "    \n",
    "    sql_Seq= f\"select Gene.ID_genebank from Gene where Gene.length <{num}\"\n",
    "    Cursor.execute(sql_Seq)\n",
    "    for row in Cursor:\n",
    "        just_seq.append(str(row))\n",
    "        \n",
    "except mysql.connector.Error as e:\n",
    "    print(\"Erro na escrita na base de dados: {}\".format(e) )    \n",
    "    \n",
    "print (Seq_and_length)\n",
    "\n",
    "if just_seq == []:\n",
    "        print()\n",
    "        print('nenhuma seq inferior a esse tamanho, pesquise de novo.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "9f33dd74",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "escolha a primeira seq com que quer trabalhar: 3\n",
      "tipo de molécula com que está a trabalhar: Protein\n",
      "[Seq('GAGTGGAGATGGCGGAGCTGT'), Seq('CAACACGGGAAACCTCACCC'), Seq('GGGATGGAAGATGGCGGCCTA'), Seq('GGTGGCTAGGGCCGCGCTGGT'), Seq('GTTTGGTGAACTTCTGCCTA')]\n",
      "GGTGGCTAGGGCCGCGCTGGT\n"
     ]
    }
   ],
   "source": [
    "list_seq=[]\n",
    "choose= input('escolha a primeira seq com que quer trabalhar: ')\n",
    "inter= int(choose)\n",
    "extr= ', '.join(just_seq)\n",
    "h= extr.replace(\"(\",'')\n",
    "hh= h.replace(\",)\",'')\n",
    "ID_al= hh.split(', ')\n",
    "select= ID_al[inter]\n",
    "list_seq.append(select)\n",
    "\n",
    "# choose2= input('escolha a segunda seq com que quer trabalhar: ')\n",
    "# inter2= int(choose2)\n",
    "# extr2= ', '.join(just_seq)\n",
    "# h2= extr2.replace(\"(\",'')\n",
    "# hh2= h2.replace(\",)\",'')\n",
    "# ID_al2= hh2.split(', ')\n",
    "# select2= ID_al2[inter2]\n",
    "# list_seq.append(select2)\n",
    "\n",
    "limpar_aspas_all= ', '.join(ID_al)\n",
    "retirar_all= limpar_aspas_all.replace(\"'\",\"\")\n",
    "otput_final_all= retirar_all.split(', ')\n",
    "\n",
    "limpar_aspas= ', '.join(list_seq)\n",
    "retirar= limpar_aspas.replace(\"'\",\"\")\n",
    "otput_final= retirar.split(', ')\n",
    "\n",
    "\n",
    "#cria lista com todos as seq com o tamanho máximo definido:\n",
    "todas_as_seq=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist= otput_final_all\n",
    "handle = Entrez.efetch(db=database, id=idlist, rettype=\"gb\") \n",
    "records = list(SeqIO.parse(handle,\"gb\"))\n",
    "handle.close()\n",
    "for info in records:\n",
    "    seqs= (info.seq)\n",
    "    todas_as_seq.append(seqs)\n",
    "    \n",
    "choose2= input('tipo de molécula com que está a trabalhar: ')    \n",
    "#cria lista só com as duas seq escolhidas para o alinhamento \n",
    "seq_for_al=[]\n",
    "database = 'nucleotide'\n",
    "email= 'rodrigoce9@gmail.com'\n",
    "idlist1= otput_final[0]\n",
    "# idlist2= otput_final[1]\n",
    "handle1 = Entrez.efetch(db=database, id=idlist1, rettype=\"gb\") \n",
    "records1 = SeqIO.read(handle1,\"gb\")\n",
    "handle1.close()\n",
    "s1_=records1.seq\n",
    "# handle2 = Entrez.efetch(db=database, id=idlist2, rettype=\"gb\") \n",
    "# records2 = SeqIO.read(handle2,\"gb\")\n",
    "# handle2.close()\n",
    "# s2_=records2.seq\n",
    "print(todas_as_seq)\n",
    "print(s1_)\n",
    "# print(s2_)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7979ffe5",
   "metadata": {},
   "source": [
    "__Goal__ Implementation of sequence manipulation algorithms: transcription, translation, ORFs, extraction of all potential proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "e37ecbb6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Don´t exist'"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#function that allows us to associate a codon with an amino acid- Human\n",
    "def translate_codon(cod):\n",
    "    \"\"\"Translates a codon into an aminoacid using an internal dictionary with the standard genetic code.\"\"\"\n",
    "    tc = {\"GCT\":\"A\", \"GCC\":\"A\", \"GCA\":\"A\", \"GCG\":\"A\", \n",
    "      \"TGT\":\"C\", \"TGC\":\"C\",\n",
    "      \"GAT\":\"D\", \"GAC\":\"D\",\n",
    "      \"GAA\":\"E\", \"GAG\":\"E\",\n",
    "      \"TTT\":\"F\", \"TTC\":\"F\",\n",
    "      \"GGT\":\"G\", \"GGC\":\"G\", \"GGA\":\"G\", \"GGG\":\"G\",\n",
    "      \"CAT\":\"H\", \"CAC\":\"H\",\n",
    "      \"ATA\":\"I\", \"ATT\":\"I\", \"ATC\":\"I\",\n",
    "      \"AAA\":\"K\", \"AAG\":\"K\",\n",
    "      \"TTA\":\"L\", \"TTG\":\"L\", \"CTT\":\"L\", \"CTC\":\"L\", \"CTA\":\"L\", \"CTG\":\"L\",\n",
    "      \"ATG\":\"M\", \"AAT\":\"N\", \"AAC\":\"N\",\n",
    "      \"CCT\":\"P\", \"CCC\":\"P\", \"CCA\":\"P\", \"CCG\":\"P\",\n",
    "      \"CAA\":\"Q\", \"CAG\":\"Q\",\n",
    "      \"CGT\":\"R\", \"CGC\":\"R\", \"CGA\":\"R\", \"CGG\":\"R\", \"AGA\":\"R\", \"AGG\":\"R\",\n",
    "      \"TCT\":\"S\", \"TCC\":\"S\", \"TCA\":\"S\", \"TCG\":\"S\", \"AGT\":\"S\", \"AGC\":\"S\",\n",
    "      \"ACT\":\"T\", \"ACC\":\"T\", \"ACA\":\"T\", \"ACG\":\"T\",\n",
    "      \"GTT\":\"V\", \"GTC\":\"V\", \"GTA\":\"V\", \"GTG\":\"V\",\n",
    "      \"TGG\":\"W\",\n",
    "      \"TAT\":\"Y\", \"TAC\":\"Y\",\n",
    "      \"TAA\":\"_\", \"TAG\":\"_\", \"TGA\":\"_\"}\n",
    "    if cod in tc: return tc[cod] \n",
    "    else: return \"Don´t exist\"\n",
    "translate_codon(\"xbt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "9c9ab58a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#talvez colocar mais possibilidade porque eu penso que a tradução muda para cada organismo??"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "afd327cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence 1 is Protein? False\n",
      "CAACACGGGAAACCTCACCC\n",
      "only do transcription with DNA\n",
      "only do reverse complement with DNA\n",
      "QHGKPH\n",
      "only want DNA orf\n",
      "just work from dna orf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#a class that allows to accomplish the objective\n",
    "class sequence_info:\n",
    "    #\"building block\", where you always have to have the sequence you are analyzing and the type of molecule: dna, rna, protein\n",
    "    def __init__(self,sequence:s1_, sequence_type:choose2):\n",
    "        self.sequence=sequence.upper() #uppercase sequence as a universal rule\n",
    "        self.sequence_type=sequence_type\n",
    "    def __length__(self):\n",
    "        return len(self.sequence)\n",
    "    def __getitem__(self,n):\n",
    "        return self.sequence[n]\n",
    "    def __getslice__(self,i,j):\n",
    "        return self.sequence[i:j]\n",
    "    def __str__(self):\n",
    "        return self.sequence\n",
    "    def sequence_type(self):\n",
    "        return self.sequence_type\n",
    "    def letters(self):\n",
    "        #each letter possible to have for dna, rna , protein for later to be able to validate the information about the type of molecule corresponding to the sequence\n",
    "        if (self.sequence_type==\"DNA\"): return \"ACGT\"\n",
    "        elif(self.sequence_type==\"RNA\"):return \"ACGU\"\n",
    "        elif(self.sequence_type==\"PROTEIN\"): return \"ACDEFGHIKLMNPQRSTVWY\"\n",
    "        else: return \"ERROR\"\n",
    "    def validate(self): #validates information about the type of molecule associated with the sequence\n",
    "        letras=self.letters()\n",
    "        res=True\n",
    "        i=0\n",
    "        while i<len(self.sequence) and res:\n",
    "            if self.sequence[i] not in letras: res = False\n",
    "            else: i +=1\n",
    "        return res\n",
    "    def contagem(self):#know which nucleotides are the most frequent amino acids\n",
    "        contagem={}\n",
    "        for x in self.sequence:\n",
    "            contagem[x]=contagem.get(x,0)+1\n",
    "            for c in sorte(contagem):\n",
    "                return c, contagem [c]       \n",
    "    def reverse_complement(self):#falta comprovar isto\n",
    "        if (self.sequence_type != \"DNA\"):return \"only do reverse complement with DNA\"\n",
    "        t=self.sequence.replace(\"T\",\"a\")\n",
    "        g=t.replace(\"G\",\"c\")\n",
    "        c=g.replace(\"C\",\"g\")\n",
    "        a=c.replace(\"A\",\"t\")\n",
    "        reverse_comp=a[::-1]\n",
    "        return reverse_comp.upper()    \n",
    "    def transcription(self):\n",
    "        if (self.sequence_type==\"DNA\"): return sequence_info(self.sequence.replace(\"T\",\"U\"),\"RNA\")\n",
    "        else:return \"only do transcription with DNA\"\n",
    "    def translate(self):\n",
    "        #if (self.sequence_type != \"DNA\"): return \"only want to do translation with DNA\"\n",
    "        sequence_aa=\"\"\n",
    "        for x in range(0, len(self.sequence)-2,3):#3 by 3\n",
    "            codon=self.sequence[x:x+3]\n",
    "            sequence_aa += translate_codon(codon)\n",
    "        return sequence_aa\n",
    "    def get_orfs_DNA(self): #6 possible orfs\n",
    "        if (self.sequence_type != \"DNA\"): return \"only want DNA orf\"\n",
    "        inverse_sequence=self.sequence[::-1]\n",
    "        return \",\".join([self.sequence[p:]for p in range(3)] + [inverse_sequence[p:] for p in range(3)]) \n",
    "    def get_orfs_Proteins(self): # translation of the 6 possible dna orfs\n",
    "        if (self.sequence_type != \"DNA\"): return \"just work from dna orf\"\n",
    "        inverse_sequence=self.sequence[::-1]\n",
    "        orfs_protein=[]\n",
    "        for p in range(3):\n",
    "            orfs_protein.append(sequence_info(self.sequence[p:],self.sequence_type).translate())\n",
    "            orfs_protein.append(sequence_info(inverse_sequence[p:],self.sequence_type).translate())\n",
    "        return \",\".join(orfs_protein)\n",
    "    def protsInAA(self):#gives all possible proteins of an amino acid sequence( from M to _)\n",
    "        seq_aa = self.translate()\n",
    "        prots_act = []\n",
    "        prots = []\n",
    "        for aa in seq_aa:\n",
    "            if aa == '_':\n",
    "                if prots_act:\n",
    "                    for p in prots_act:\n",
    "                        prots.append(p)\n",
    "                    prots_act = []\n",
    "            else:\n",
    "                if aa == 'M':\n",
    "                    prots_act.append('')\n",
    "                for i in range(len(prots_act)):\n",
    "                    prots_act[i] += aa\n",
    "        return \",\".join(prots) \n",
    " #if name=main é para conseguir importar direitinho os nodulos ou seja quanto estou neste ficheiro o name =main e ele faz a classe se for para outro ficheiro e importar a classe ele faz na mesma a classe sem repetir tufo o que aqui tenho\n",
    "# https://www.youtube.com/watch?v=3dHyS1W4TIE\n",
    "if __name__ == \"__main__\":\n",
    "    s1_=\"CAACACGGGAAACCTCACCC\"\n",
    "    s1 = sequence_info(s1_,choose2)\n",
    "    s2 = sequence_info(\"MKVVLSVQERSVVSLL_\", \"PROTEIN\")\n",
    "    s3 = sequence_info(\"ATTTTBTT\",\"PROTEIN\")\n",
    "    print(\"sequence 1 is {}? {}\".format(s1.sequence_type,s1.validate()))\n",
    "    print(s1)\n",
    "    s3 = s1.transcription()\n",
    "    print(s3)\n",
    "    s4 = s1.reverse_complement()\n",
    "    print(s4)\n",
    "    s5= s1.translate()\n",
    "    print(s5)\n",
    "    s6=s1.get_orfs_DNA()\n",
    "    print(s6)\n",
    "    s8=s1.get_orfs_Proteins()\n",
    "    print(s8)\n",
    "    s9=s1.protsInAA()\n",
    "    print(s9)\n",
    "    \n",
    "\n",
    "    \n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
